<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>CRISPR-AnalyzeR Report</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="CRISPR-AnalyzeR Report">
  <meta name="generator" content="bookdown 0.3 and GitBook 2.6.7">

  <meta property="og:title" content="CRISPR-AnalyzeR Report" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="CRISPR-AnalyzeR Report" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="sample-comparison.html">
<link rel="next" href="scatter-plot-matrices.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-0.8/htmlwidgets.js"></script>
<script src="libs/datatables-binding-0.2/datatables.js"></script>
<link href="libs/dt-core-1.10.12/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.10.12/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.10.12/js/jquery.dataTables.min.js"></script>
<script src="libs/jszip-1.10.12/jszip.min.js"></script>
<script src="libs/pdfmake-1.10.12/pdfmake.min.js"></script>
<script src="libs/pdfmake-1.10.12/vfs_fonts.js"></script>
<link href="libs/dt-ext-buttons-1.10.12/css/buttons.dataTables.min.css" rel="stylesheet" />
<script src="libs/dt-ext-buttons-1.10.12/js/dataTables.buttons.min.js"></script>
<script src="libs/dt-ext-buttons-1.10.12/js/buttons.flash.min.js"></script>
<script src="libs/dt-ext-buttons-1.10.12/js/buttons.html5.min.js"></script>
<script src="libs/dt-ext-buttons-1.10.12/js/buttons.colVis.min.js"></script>
<script src="libs/dt-ext-buttons-1.10.12/js/buttons.print.min.js"></script>
<script src="libs/proj4js-2.3.15/proj4.js"></script>
<link href="libs/highcharts-5.0.6/css/motion.css" rel="stylesheet" />
<script src="libs/highcharts-5.0.6/highstock.js"></script>
<script src="libs/highcharts-5.0.6/highcharts-3d.js"></script>
<script src="libs/highcharts-5.0.6/highcharts-more.js"></script>
<script src="libs/highcharts-5.0.6/modules/annotations.js"></script>
<script src="libs/highcharts-5.0.6/modules/broken-axis.js"></script>
<script src="libs/highcharts-5.0.6/modules/data.js"></script>
<script src="libs/highcharts-5.0.6/modules/drilldown.js"></script>
<script src="libs/highcharts-5.0.6/modules/exporting.js"></script>
<script src="libs/highcharts-5.0.6/modules/funnel.js"></script>
<script src="libs/highcharts-5.0.6/modules/heatmap.js"></script>
<script src="libs/highcharts-5.0.6/modules/map.js"></script>
<script src="libs/highcharts-5.0.6/modules/no-data-to-display.js"></script>
<script src="libs/highcharts-5.0.6/modules/offline-exporting.js"></script>
<script src="libs/highcharts-5.0.6/modules/solid-gauge.js"></script>
<script src="libs/highcharts-5.0.6/modules/treemap.js"></script>
<script src="libs/highcharts-5.0.6/plugins/annotations.js"></script>
<script src="libs/highcharts-5.0.6/plugins/draggable-legend.js"></script>
<script src="libs/highcharts-5.0.6/plugins/draggable-points.js"></script>
<script src="libs/highcharts-5.0.6/plugins/export-csv.js"></script>
<script src="libs/highcharts-5.0.6/plugins/grouped-categories.js"></script>
<script src="libs/highcharts-5.0.6/plugins/motion.js"></script>
<script src="libs/highcharts-5.0.6/plugins/pattern-fill-v2.js"></script>
<script src="libs/highcharts-5.0.6/plugins/tooltip-delay.js"></script>
<script src="libs/highcharts-5.0.6/custom/reset.js"></script>
<script src="libs/highcharts-5.0.6/custom/symbols-extra.js"></script>
<script src="libs/highcharts-5.0.6/custom/text-symbols.js"></script>
<link href="libs/fontawesome-4.5.0/font-awesome.min.css" rel="stylesheet" />
<link href="libs/htmlwdgtgrid-1/htmlwdgtgrid.css" rel="stylesheet" />
<script src="libs/highchart-binding-0.5.0/highchart.js"></script>



</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="screen-information.html"><a href="screen-information.html"><i class="fa fa-check"></i><b>1</b> Screen Information</a><ul>
<li class="chapter" data-level="1.1" data-path="experiment.html"><a href="experiment.html"><i class="fa fa-check"></i><b>1.1</b> Experiment</a><ul>
<li class="chapter" data-level="1.1.1" data-path="experiment.html"><a href="experiment.html#scope"><i class="fa fa-check"></i><b>1.1.1</b> Scope</a></li>
<li class="chapter" data-level="1.1.2" data-path="experiment.html"><a href="experiment.html#procedure"><i class="fa fa-check"></i><b>1.1.2</b> Procedure</a></li>
<li class="chapter" data-level="1.1.3" data-path="experiment.html"><a href="experiment.html#comments"><i class="fa fa-check"></i><b>1.1.3</b> Comments</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="data.html"><a href="data.html"><i class="fa fa-check"></i><b>1.2</b> Data</a><ul>
<li class="chapter" data-level="1.2.1" data-path="data.html"><a href="data.html#files"><i class="fa fa-check"></i><b>1.2.1</b> Files</a></li>
<li class="chapter" data-level="1.2.2" data-path="data.html"><a href="data.html#experimental-design"><i class="fa fa-check"></i><b>1.2.2</b> Experimental Design</a></li>
<li class="chapter" data-level="1.2.3" data-path="data.html"><a href="data.html#analysis-parameters"><i class="fa fa-check"></i><b>1.2.3</b> Analysis Parameters</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>I Screen Quality</b></span></li>
<li class="chapter" data-level="2" data-path="data-review.html"><a href="data-review.html"><i class="fa fa-check"></i><b>2</b> Data Review</a></li>
<li class="chapter" data-level="3" data-path="fastq-quality.html"><a href="fastq-quality.html"><i class="fa fa-check"></i><b>3</b> FASTQ Quality</a><ul>
<li class="chapter" data-level="3.1" data-path="cycle-specific-gc-content.html"><a href="cycle-specific-gc-content.html"><i class="fa fa-check"></i><b>3.1</b> Cycle specific GC Content</a></li>
<li class="chapter" data-level="3.2" data-path="cycle-specific-quality-distribution.html"><a href="cycle-specific-quality-distribution.html"><i class="fa fa-check"></i><b>3.2</b> Cycle specific Quality Distribution</a></li>
<li class="chapter" data-level="3.3" data-path="cycle-specific-average-quality.html"><a href="cycle-specific-average-quality.html"><i class="fa fa-check"></i><b>3.3</b> Cycle specific Average Quality</a></li>
<li class="chapter" data-level="3.4" data-path="cycle-specific-base-call-proportion.html"><a href="cycle-specific-base-call-proportion.html"><i class="fa fa-check"></i><b>3.4</b> Cycle specific Base Call Proportion</a></li>
<li class="chapter" data-level="3.5" data-path="read-frequency.html"><a href="read-frequency.html"><i class="fa fa-check"></i><b>3.5</b> Read Frequency</a></li>
<li class="chapter" data-level="3.6" data-path="read-width.html"><a href="read-width.html"><i class="fa fa-check"></i><b>3.6</b> Read Width</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="screen-readcount.html"><a href="screen-readcount.html"><i class="fa fa-check"></i><b>4</b> Screen Readcount</a><ul>
<li class="chapter" data-level="4.1" data-path="overview.html"><a href="overview.html"><i class="fa fa-check"></i><b>4.1</b> Overview</a><ul>
<li class="chapter" data-level="4.1.1" data-path="overview.html"><a href="overview.html#sgrna-readcount-statistics"><i class="fa fa-check"></i><b>4.1.1</b> sgRNA Readcount Statistics</a></li>
<li class="chapter" data-level="4.1.2" data-path="overview.html"><a href="overview.html#gene-readcount-statistics"><i class="fa fa-check"></i><b>4.1.2</b> Gene Readcount Statistics</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="readcount-distribution.html"><a href="readcount-distribution.html"><i class="fa fa-check"></i><b>4.2</b> Readcount Distribution</a><ul>
<li class="chapter" data-level="4.2.1" data-path="readcount-distribution.html"><a href="readcount-distribution.html#pseudodensity"><i class="fa fa-check"></i><b>4.2.1</b> Pseudodensity</a></li>
<li class="chapter" data-level="4.2.2" data-path="readcount-distribution.html"><a href="readcount-distribution.html#cumulative-frequency-for-gene-read-counts"><i class="fa fa-check"></i><b>4.2.2</b> Cumulative Frequency for Gene Read Counts</a></li>
<li class="chapter" data-level="4.2.3" data-path="readcount-distribution.html"><a href="readcount-distribution.html#cumulative-frequency-for-sgrna-read-counts"><i class="fa fa-check"></i><b>4.2.3</b> Cumulative Frequency for sgRNA Read Counts</a></li>
<li class="chapter" data-level="4.2.4" data-path="readcount-distribution.html"><a href="readcount-distribution.html#boxplots"><i class="fa fa-check"></i><b>4.2.4</b> Boxplots</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="controls.html"><a href="controls.html"><i class="fa fa-check"></i><b>4.3</b> Controls</a><ul>
<li class="chapter" data-level="4.3.1" data-path="controls.html"><a href="controls.html#positive-controls"><i class="fa fa-check"></i><b>4.3.1</b> Positive Controls</a></li>
<li class="chapter" data-level="4.3.2" data-path="controls.html"><a href="controls.html#non-targeting-controls"><i class="fa fa-check"></i><b>4.3.2</b> Non-targeting Controls</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="readcount-depth.html"><a href="readcount-depth.html"><i class="fa fa-check"></i><b>4.4</b> Readcount Depth</a><ul>
<li class="chapter" data-level="4.4.1" data-path="readcount-depth.html"><a href="readcount-depth.html#pbs_replicate1"><i class="fa fa-check"></i><b>4.4.1</b> PBS_Replicate1</a></li>
<li class="chapter" data-level="4.4.2" data-path="readcount-depth.html"><a href="readcount-depth.html#pbs_replicate2"><i class="fa fa-check"></i><b>4.4.2</b> PBS_Replicate2</a></li>
<li class="chapter" data-level="4.4.3" data-path="readcount-depth.html"><a href="readcount-depth.html#trail_replicate2"><i class="fa fa-check"></i><b>4.4.3</b> TRAIL_Replicate2</a></li>
<li class="chapter" data-level="4.4.4" data-path="readcount-depth.html"><a href="readcount-depth.html#trail_replicate1"><i class="fa fa-check"></i><b>4.4.4</b> TRAIL_Replicate1</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="readcounts.html"><a href="readcounts.html"><i class="fa fa-check"></i><b>4.5</b> Readcounts</a><ul>
<li class="chapter" data-level="4.5.1" data-path="readcounts.html"><a href="readcounts.html#pbs_replicate1-1"><i class="fa fa-check"></i><b>4.5.1</b> PBS_Replicate1</a></li>
<li class="chapter" data-level="4.5.2" data-path="readcounts.html"><a href="readcounts.html#pbs_replicate2-1"><i class="fa fa-check"></i><b>4.5.2</b> PBS_Replicate2</a></li>
<li class="chapter" data-level="4.5.3" data-path="readcounts.html"><a href="readcounts.html#trail_replicate2-1"><i class="fa fa-check"></i><b>4.5.3</b> TRAIL_Replicate2</a></li>
<li class="chapter" data-level="4.5.4" data-path="readcounts.html"><a href="readcounts.html#trail_replicate1-1"><i class="fa fa-check"></i><b>4.5.4</b> TRAIL_Replicate1</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="sgrna-coverage.html"><a href="sgrna-coverage.html"><i class="fa fa-check"></i><b>5</b> sgRNA Coverage</a><ul>
<li class="chapter" data-level="5.1" data-path="missing-sgrnas-and-genes.html"><a href="missing-sgrnas-and-genes.html"><i class="fa fa-check"></i><b>5.1</b> Missing sgRNAs and Genes</a><ul>
<li class="chapter" data-level="5.1.1" data-path="missing-sgrnas-and-genes.html"><a href="missing-sgrnas-and-genes.html#missing-genes"><i class="fa fa-check"></i><b>5.1.1</b> Missing Genes</a></li>
<li class="chapter" data-level="5.1.2" data-path="missing-sgrnas-and-genes.html"><a href="missing-sgrnas-and-genes.html#missing-sgrnas"><i class="fa fa-check"></i><b>5.1.2</b> Missing sgRNAs</a></li>
<li class="chapter" data-level="5.1.3" data-path="missing-sgrnas-and-genes.html"><a href="missing-sgrnas-and-genes.html#missing-non-targeting-sgrnas"><i class="fa fa-check"></i><b>5.1.3</b> % Missing Non-targeting sgRNAs</a></li>
<li class="chapter" data-level="5.1.4" data-path="missing-sgrnas-and-genes.html"><a href="missing-sgrnas-and-genes.html#missing-postive-sgrnas"><i class="fa fa-check"></i><b>5.1.4</b> % Missing Postive sgRNAs</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="designs-per-gene.html"><a href="designs-per-gene.html"><i class="fa fa-check"></i><b>5.2</b> Designs per Gene</a><ul>
<li class="chapter" data-level="5.2.1" data-path="designs-per-gene.html"><a href="designs-per-gene.html#pbs_replicate1-2"><i class="fa fa-check"></i><b>5.2.1</b> PBS_Replicate1</a></li>
<li class="chapter" data-level="5.2.2" data-path="designs-per-gene.html"><a href="designs-per-gene.html#pbs_replicate2-2"><i class="fa fa-check"></i><b>5.2.2</b> PBS_Replicate2</a></li>
<li class="chapter" data-level="5.2.3" data-path="designs-per-gene.html"><a href="designs-per-gene.html#trail_replicate2-2"><i class="fa fa-check"></i><b>5.2.3</b> TRAIL_Replicate2</a></li>
<li class="chapter" data-level="5.2.4" data-path="designs-per-gene.html"><a href="designs-per-gene.html#trail_replicate1-2"><i class="fa fa-check"></i><b>5.2.4</b> TRAIL_Replicate1</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sample-comparison.html"><a href="sample-comparison.html"><i class="fa fa-check"></i><b>6</b> Sample Comparison</a><ul>
<li class="chapter" data-level="6.1" data-path="overview-1.html"><a href="overview-1.html"><i class="fa fa-check"></i><b>6.1</b> Overview</a><ul>
<li class="chapter" data-level="6.1.1" data-path="overview-1.html"><a href="overview-1.html#foldchange-of-gene-effects"><i class="fa fa-check"></i><b>6.1.1</b> Foldchange of Gene Effects</a></li>
<li class="chapter" data-level="6.1.2" data-path="overview-1.html"><a href="overview-1.html#z-ratio-of-gene-effects"><i class="fa fa-check"></i><b>6.1.2</b> Z-Ratio of Gene Effects</a></li>
<li class="chapter" data-level="6.1.3" data-path="overview-1.html"><a href="overview-1.html#foldchange-of-sgrna-effects"><i class="fa fa-check"></i><b>6.1.3</b> Foldchange of sgRNA Effects</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="scatter-plot-matrices.html"><a href="scatter-plot-matrices.html"><i class="fa fa-check"></i><b>6.2</b> Scatter Plot Matrices</a><ul>
<li class="chapter" data-level="6.2.1" data-path="scatter-plot-matrices.html"><a href="scatter-plot-matrices.html#scatter-plot-matrix-1"><i class="fa fa-check"></i><b>6.2.1</b> Scatter Plot Matrix 1</a></li>
<li class="chapter" data-level="6.2.2" data-path="scatter-plot-matrices.html"><a href="scatter-plot-matrices.html#scatter-plot-matrix-2"><i class="fa fa-check"></i><b>6.2.2</b> Scatter Plot Matrix 2</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="single-scatter-plots.html"><a href="single-scatter-plots.html"><i class="fa fa-check"></i><b>6.3</b> Single Scatter Plots</a><ul>
<li class="chapter" data-level="6.3.1" data-path="single-scatter-plots.html"><a href="single-scatter-plots.html#scatterplot-1"><i class="fa fa-check"></i><b>6.3.1</b> Scatterplot 1</a></li>
<li class="chapter" data-level="6.3.2" data-path="single-scatter-plots.html"><a href="single-scatter-plots.html#scatterplot-2"><i class="fa fa-check"></i><b>6.3.2</b> Scatterplot 2</a></li>
<li class="chapter" data-level="6.3.3" data-path="single-scatter-plots.html"><a href="single-scatter-plots.html#scatterplot-3"><i class="fa fa-check"></i><b>6.3.3</b> Scatterplot 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="pca.html"><a href="pca.html"><i class="fa fa-check"></i><b>7</b> PCA</a></li>
<li class="part"><span><b>II Hit Calling</b></span></li>
<li class="chapter" data-level="8" data-path="gene-ranking.html"><a href="gene-ranking.html"><i class="fa fa-check"></i><b>8</b> Gene Ranking</a><ul>
<li class="chapter" data-level="8.1" data-path="wilcox.html"><a href="wilcox.html"><i class="fa fa-check"></i><b>8.1</b> Wilcox</a><ul>
<li class="chapter" data-level="8.1.1" data-path="wilcox.html"><a href="wilcox.html#p-value-distribution"><i class="fa fa-check"></i><b>8.1.1</b> P Value Distribution</a></li>
<li class="chapter" data-level="8.1.2" data-path="wilcox.html"><a href="wilcox.html#ranked-p-values"><i class="fa fa-check"></i><b>8.1.2</b> Ranked P Values</a></li>
<li class="chapter" data-level="8.1.3" data-path="wilcox.html"><a href="wilcox.html#p-value-vs-lfc"><i class="fa fa-check"></i><b>8.1.3</b> P Value vs LFC</a></li>
<li class="chapter" data-level="8.1.4" data-path="wilcox.html"><a href="wilcox.html#data-1"><i class="fa fa-check"></i><b>8.1.4</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="deseq2.html"><a href="deseq2.html"><i class="fa fa-check"></i><b>8.2</b> DESeq2</a><ul>
<li class="chapter" data-level="8.2.1" data-path="deseq2.html"><a href="deseq2.html#p-value-distribution-1"><i class="fa fa-check"></i><b>8.2.1</b> P Value Distribution</a></li>
<li class="chapter" data-level="8.2.2" data-path="deseq2.html"><a href="deseq2.html#ranked-p-values-1"><i class="fa fa-check"></i><b>8.2.2</b> Ranked P Values</a></li>
<li class="chapter" data-level="8.2.3" data-path="deseq2.html"><a href="deseq2.html#p-value-vs-lfc-1"><i class="fa fa-check"></i><b>8.2.3</b> P Value vs LFC</a></li>
<li class="chapter" data-level="8.2.4" data-path="deseq2.html"><a href="deseq2.html#data-2"><i class="fa fa-check"></i><b>8.2.4</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="mageck.html"><a href="mageck.html"><i class="fa fa-check"></i><b>8.3</b> MAGeCK</a><ul>
<li class="chapter" data-level="8.3.1" data-path="mageck.html"><a href="mageck.html#p-value-distribution-2"><i class="fa fa-check"></i><b>8.3.1</b> P Value Distribution</a></li>
<li class="chapter" data-level="8.3.2" data-path="mageck.html"><a href="mageck.html#ranked-p-values-2"><i class="fa fa-check"></i><b>8.3.2</b> Ranked P Values</a></li>
<li class="chapter" data-level="8.3.3" data-path="mageck.html"><a href="mageck.html#p-value-vs-lfc-2"><i class="fa fa-check"></i><b>8.3.3</b> P Value vs LFC</a></li>
<li class="chapter" data-level="8.3.4" data-path="mageck.html"><a href="mageck.html#data-3"><i class="fa fa-check"></i><b>8.3.4</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="sgrsea.html"><a href="sgrsea.html"><i class="fa fa-check"></i><b>8.4</b> sgRSEA</a><ul>
<li class="chapter" data-level="8.4.1" data-path="sgrsea.html"><a href="sgrsea.html#p-value-distribution-3"><i class="fa fa-check"></i><b>8.4.1</b> P Value Distribution</a></li>
<li class="chapter" data-level="8.4.2" data-path="sgrsea.html"><a href="sgrsea.html#ranked-p-values-3"><i class="fa fa-check"></i><b>8.4.2</b> Ranked P Values</a></li>
<li class="chapter" data-level="8.4.3" data-path="sgrsea.html"><a href="sgrsea.html#p-value-vs-lfc-3"><i class="fa fa-check"></i><b>8.4.3</b> P Value vs LFC</a></li>
<li class="chapter" data-level="8.4.4" data-path="sgrsea.html"><a href="sgrsea.html#data-4"><i class="fa fa-check"></i><b>8.4.4</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="edger.html"><a href="edger.html"><i class="fa fa-check"></i><b>8.5</b> EdgeR</a><ul>
<li class="chapter" data-level="8.5.1" data-path="edger.html"><a href="edger.html#p-value-distribution-4"><i class="fa fa-check"></i><b>8.5.1</b> P Value Distribution</a></li>
<li class="chapter" data-level="8.5.2" data-path="edger.html"><a href="edger.html#ranked-p-values-4"><i class="fa fa-check"></i><b>8.5.2</b> Ranked P Values</a></li>
<li class="chapter" data-level="8.5.3" data-path="edger.html"><a href="edger.html#p-value-vs-lfc-4"><i class="fa fa-check"></i><b>8.5.3</b> P Value vs LFC</a></li>
<li class="chapter" data-level="8.5.4" data-path="edger.html"><a href="edger.html#data-5"><i class="fa fa-check"></i><b>8.5.4</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="z-ratio.html"><a href="z-ratio.html"><i class="fa fa-check"></i><b>8.6</b> Z Ratio</a><ul>
<li class="chapter" data-level="8.6.1" data-path="z-ratio.html"><a href="z-ratio.html#z-ratio-distribution"><i class="fa fa-check"></i><b>8.6.1</b> Z Ratio Distribution</a></li>
<li class="chapter" data-level="8.6.2" data-path="z-ratio.html"><a href="z-ratio.html#ranked-z-ratios"><i class="fa fa-check"></i><b>8.6.2</b> Ranked Z Ratios</a></li>
<li class="chapter" data-level="8.6.3" data-path="z-ratio.html"><a href="z-ratio.html#z-ratio-vs-lfc"><i class="fa fa-check"></i><b>8.6.3</b> Z Ratio vs LFC</a></li>
<li class="chapter" data-level="8.6.4" data-path="z-ratio.html"><a href="z-ratio.html#data-6"><i class="fa fa-check"></i><b>8.6.4</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="bagel.html"><a href="bagel.html"><i class="fa fa-check"></i><b>8.7</b> BAGEL</a><ul>
<li class="chapter" data-level="8.7.1" data-path="bagel.html"><a href="bagel.html#log2-bayes-factor-distribution"><i class="fa fa-check"></i><b>8.7.1</b> Log2 Bayes Factor Distribution</a></li>
<li class="chapter" data-level="8.7.2" data-path="bagel.html"><a href="bagel.html#ranked-by-log2-of-bayes-factor"><i class="fa fa-check"></i><b>8.7.2</b> Ranked by log2 of Bayes Factor</a></li>
<li class="chapter" data-level="8.7.3" data-path="bagel.html"><a href="bagel.html#data-7"><i class="fa fa-check"></i><b>8.7.3</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="screenbeam.html"><a href="screenbeam.html"><i class="fa fa-check"></i><b>8.8</b> ScreenBEAM</a><ul>
<li class="chapter" data-level="8.8.1" data-path="screenbeam.html"><a href="screenbeam.html#p-value-distribution-5"><i class="fa fa-check"></i><b>8.8.1</b> P-Value Distribution</a></li>
<li class="chapter" data-level="8.8.2" data-path="screenbeam.html"><a href="screenbeam.html#ranked-by-fdr"><i class="fa fa-check"></i><b>8.8.2</b> Ranked by FDR</a></li>
<li class="chapter" data-level="8.8.3" data-path="screenbeam.html"><a href="screenbeam.html#data-8"><i class="fa fa-check"></i><b>8.8.3</b> Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="summary.html"><a href="summary.html"><i class="fa fa-check"></i><b>9</b> Summary</a><ul>
<li class="chapter" data-level="9.1" data-path="enriched-11.html"><a href="enriched-11.html"><i class="fa fa-check"></i><b>9.1</b> Enriched</a></li>
<li class="chapter" data-level="9.2" data-path="depleted-11.html"><a href="depleted-11.html"><i class="fa fa-check"></i><b>9.2</b> Depleted</a></li>
</ul></li>
<li class="part"><span><b>III Hit Confirmation</b></span></li>
<li class="chapter" data-level="10" data-path="gene-overviews.html"><a href="gene-overviews.html"><i class="fa fa-check"></i><b>10</b> Gene Overviews</a><ul>
<li class="chapter" data-level="10.1" data-path="casp3-1.html"><a href="casp3-1.html"><i class="fa fa-check"></i><b>10.1</b> CASP3 (1)</a><ul>
<li class="chapter" data-level="10.1.1" data-path="casp3-1.html"><a href="casp3-1.html#general-gene-information"><i class="fa fa-check"></i><b>10.1.1</b> General Gene Information</a></li>
<li class="chapter" data-level="10.1.2" data-path="casp3-1.html"><a href="casp3-1.html#gene-model"><i class="fa fa-check"></i><b>10.1.2</b> Gene Model</a></li>
<li class="chapter" data-level="10.1.3" data-path="casp3-1.html"><a href="casp3-1.html#sgrna-model"><i class="fa fa-check"></i><b>10.1.3</b> sgRNA Model</a></li>
<li class="chapter" data-level="10.1.4" data-path="casp3-1.html"><a href="casp3-1.html#published-screens"><i class="fa fa-check"></i><b>10.1.4</b> Published Screens</a></li>
<li class="chapter" data-level="10.1.5" data-path="casp3-1.html"><a href="casp3-1.html#gene-performance"><i class="fa fa-check"></i><b>10.1.5</b> Gene Performance</a></li>
<li class="chapter" data-level="10.1.6" data-path="casp3-1.html"><a href="casp3-1.html#cosmic-mutation-database"><i class="fa fa-check"></i><b>10.1.6</b> COSMIC Mutation Database</a></li>
<li class="chapter" data-level="10.1.7" data-path="casp3-1.html"><a href="casp3-1.html#gene-ontology"><i class="fa fa-check"></i><b>10.1.7</b> Gene Ontology</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html"><i class="fa fa-check"></i><b>10.2</b> TNFRSF10A (2)</a><ul>
<li class="chapter" data-level="10.2.1" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html#general-gene-information-1"><i class="fa fa-check"></i><b>10.2.1</b> General Gene Information</a></li>
<li class="chapter" data-level="10.2.2" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html#gene-model-1"><i class="fa fa-check"></i><b>10.2.2</b> Gene Model</a></li>
<li class="chapter" data-level="10.2.3" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html#sgrna-model-1"><i class="fa fa-check"></i><b>10.2.3</b> sgRNA Model</a></li>
<li class="chapter" data-level="10.2.4" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html#published-screens-1"><i class="fa fa-check"></i><b>10.2.4</b> Published Screens</a></li>
<li class="chapter" data-level="10.2.5" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html#gene-performance-1"><i class="fa fa-check"></i><b>10.2.5</b> Gene Performance</a></li>
<li class="chapter" data-level="10.2.6" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html#cosmic-mutation-database-1"><i class="fa fa-check"></i><b>10.2.6</b> COSMIC Mutation Database</a></li>
<li class="chapter" data-level="10.2.7" data-path="tnfrsf10a-2.html"><a href="tnfrsf10a-2.html#gene-ontology-1"><i class="fa fa-check"></i><b>10.2.7</b> Gene Ontology</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="casp8-3.html"><a href="casp8-3.html"><i class="fa fa-check"></i><b>10.3</b> CASP8 (3)</a><ul>
<li class="chapter" data-level="10.3.1" data-path="casp8-3.html"><a href="casp8-3.html#general-gene-information-2"><i class="fa fa-check"></i><b>10.3.1</b> General Gene Information</a></li>
<li class="chapter" data-level="10.3.2" data-path="casp8-3.html"><a href="casp8-3.html#gene-model-2"><i class="fa fa-check"></i><b>10.3.2</b> Gene Model</a></li>
<li class="chapter" data-level="10.3.3" data-path="casp8-3.html"><a href="casp8-3.html#sgrna-model-2"><i class="fa fa-check"></i><b>10.3.3</b> sgRNA Model</a></li>
<li class="chapter" data-level="10.3.4" data-path="casp8-3.html"><a href="casp8-3.html#published-screens-2"><i class="fa fa-check"></i><b>10.3.4</b> Published Screens</a></li>
<li class="chapter" data-level="10.3.5" data-path="casp8-3.html"><a href="casp8-3.html#gene-performance-2"><i class="fa fa-check"></i><b>10.3.5</b> Gene Performance</a></li>
<li class="chapter" data-level="10.3.6" data-path="casp8-3.html"><a href="casp8-3.html#cosmic-mutation-database-2"><i class="fa fa-check"></i><b>10.3.6</b> COSMIC Mutation Database</a></li>
<li class="chapter" data-level="10.3.7" data-path="casp8-3.html"><a href="casp8-3.html#gene-ontology-2"><i class="fa fa-check"></i><b>10.3.7</b> Gene Ontology</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="sgrnas.html"><a href="sgrnas.html"><i class="fa fa-check"></i><b>11</b> sgRNAs</a><ul>
<li class="chapter" data-level="11.1" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html"><i class="fa fa-check"></i><b>11.1</b> TNFRSF10B (1)</a><ul>
<li class="chapter" data-level="11.1.1" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#readcount"><i class="fa fa-check"></i><b>11.1.1</b> Readcount</a></li>
<li class="chapter" data-level="11.1.2" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#log2-foldchange"><i class="fa fa-check"></i><b>11.1.2</b> Log2 Foldchange</a></li>
<li class="chapter" data-level="11.1.3" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#genomic-binding-sites"><i class="fa fa-check"></i><b>11.1.3</b> Genomic Binding Sites</a></li>
<li class="chapter" data-level="11.1.4" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#z-score"><i class="fa fa-check"></i><b>11.1.4</b> Z-Score</a></li>
<li class="chapter" data-level="11.1.5" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#efficiency-scores"><i class="fa fa-check"></i><b>11.1.5</b> Efficiency Scores</a></li>
<li class="chapter" data-level="11.1.6" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#e-crisp-scores"><i class="fa fa-check"></i><b>11.1.6</b> E-CRISP Scores</a></li>
<li class="chapter" data-level="11.1.7" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#sgrna-sequences"><i class="fa fa-check"></i><b>11.1.7</b> sgRNA Sequences</a></li>
<li class="chapter" data-level="11.1.8" data-path="tnfrsf10b-1.html"><a href="tnfrsf10b-1.html#sgrna-binding-sites"><i class="fa fa-check"></i><b>11.1.8</b> sgRNA Binding Sites</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html"><i class="fa fa-check"></i><b>11.2</b> TNFRSF10A (2)</a><ul>
<li class="chapter" data-level="11.2.1" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#readcount-1"><i class="fa fa-check"></i><b>11.2.1</b> Readcount</a></li>
<li class="chapter" data-level="11.2.2" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#log2-foldchange-1"><i class="fa fa-check"></i><b>11.2.2</b> Log2 Foldchange</a></li>
<li class="chapter" data-level="11.2.3" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#genomic-binding-sites-1"><i class="fa fa-check"></i><b>11.2.3</b> Genomic Binding Sites</a></li>
<li class="chapter" data-level="11.2.4" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#z-score-1"><i class="fa fa-check"></i><b>11.2.4</b> Z-Score</a></li>
<li class="chapter" data-level="11.2.5" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#efficiency-scores-1"><i class="fa fa-check"></i><b>11.2.5</b> Efficiency Scores</a></li>
<li class="chapter" data-level="11.2.6" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#e-crisp-scores-1"><i class="fa fa-check"></i><b>11.2.6</b> E-CRISP Scores</a></li>
<li class="chapter" data-level="11.2.7" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#sgrna-sequences-1"><i class="fa fa-check"></i><b>11.2.7</b> sgRNA Sequences</a></li>
<li class="chapter" data-level="11.2.8" data-path="tnfrsf10a-2-1.html"><a href="tnfrsf10a-2-1.html#sgrna-binding-sites-1"><i class="fa fa-check"></i><b>11.2.8</b> sgRNA Binding Sites</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="casp3-3.html"><a href="casp3-3.html"><i class="fa fa-check"></i><b>11.3</b> CASP3 (3)</a><ul>
<li class="chapter" data-level="11.3.1" data-path="casp3-3.html"><a href="casp3-3.html#readcount-2"><i class="fa fa-check"></i><b>11.3.1</b> Readcount</a></li>
<li class="chapter" data-level="11.3.2" data-path="casp3-3.html"><a href="casp3-3.html#log2-foldchange-2"><i class="fa fa-check"></i><b>11.3.2</b> Log2 Foldchange</a></li>
<li class="chapter" data-level="11.3.3" data-path="casp3-3.html"><a href="casp3-3.html#genomic-binding-sites-2"><i class="fa fa-check"></i><b>11.3.3</b> Genomic Binding Sites</a></li>
<li class="chapter" data-level="11.3.4" data-path="casp3-3.html"><a href="casp3-3.html#z-score-2"><i class="fa fa-check"></i><b>11.3.4</b> Z-Score</a></li>
<li class="chapter" data-level="11.3.5" data-path="casp3-3.html"><a href="casp3-3.html#efficiency-scores-2"><i class="fa fa-check"></i><b>11.3.5</b> Efficiency Scores</a></li>
<li class="chapter" data-level="11.3.6" data-path="casp3-3.html"><a href="casp3-3.html#e-crisp-scores-2"><i class="fa fa-check"></i><b>11.3.6</b> E-CRISP Scores</a></li>
<li class="chapter" data-level="11.3.7" data-path="casp3-3.html"><a href="casp3-3.html#sgrna-sequences-2"><i class="fa fa-check"></i><b>11.3.7</b> sgRNA Sequences</a></li>
<li class="chapter" data-level="11.3.8" data-path="casp3-3.html"><a href="casp3-3.html#sgrna-binding-sites-2"><i class="fa fa-check"></i><b>11.3.8</b> sgRNA Binding Sites</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="fadd-4.html"><a href="fadd-4.html"><i class="fa fa-check"></i><b>11.4</b> FADD (4)</a><ul>
<li class="chapter" data-level="11.4.1" data-path="fadd-4.html"><a href="fadd-4.html#readcount-3"><i class="fa fa-check"></i><b>11.4.1</b> Readcount</a></li>
<li class="chapter" data-level="11.4.2" data-path="fadd-4.html"><a href="fadd-4.html#log2-foldchange-3"><i class="fa fa-check"></i><b>11.4.2</b> Log2 Foldchange</a></li>
<li class="chapter" data-level="11.4.3" data-path="fadd-4.html"><a href="fadd-4.html#genomic-binding-sites-3"><i class="fa fa-check"></i><b>11.4.3</b> Genomic Binding Sites</a></li>
<li class="chapter" data-level="11.4.4" data-path="fadd-4.html"><a href="fadd-4.html#z-score-3"><i class="fa fa-check"></i><b>11.4.4</b> Z-Score</a></li>
<li class="chapter" data-level="11.4.5" data-path="fadd-4.html"><a href="fadd-4.html#efficiency-scores-3"><i class="fa fa-check"></i><b>11.4.5</b> Efficiency Scores</a></li>
<li class="chapter" data-level="11.4.6" data-path="fadd-4.html"><a href="fadd-4.html#e-crisp-scores-3"><i class="fa fa-check"></i><b>11.4.6</b> E-CRISP Scores</a></li>
<li class="chapter" data-level="11.4.7" data-path="fadd-4.html"><a href="fadd-4.html#sgrna-sequences-3"><i class="fa fa-check"></i><b>11.4.7</b> sgRNA Sequences</a></li>
<li class="chapter" data-level="11.4.8" data-path="fadd-4.html"><a href="fadd-4.html#sgrna-binding-sites-3"><i class="fa fa-check"></i><b>11.4.8</b> sgRNA Binding Sites</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="casp8-5.html"><a href="casp8-5.html"><i class="fa fa-check"></i><b>11.5</b> CASP8 (5)</a><ul>
<li class="chapter" data-level="11.5.1" data-path="casp8-5.html"><a href="casp8-5.html#readcount-4"><i class="fa fa-check"></i><b>11.5.1</b> Readcount</a></li>
<li class="chapter" data-level="11.5.2" data-path="casp8-5.html"><a href="casp8-5.html#log2-foldchange-4"><i class="fa fa-check"></i><b>11.5.2</b> Log2 Foldchange</a></li>
<li class="chapter" data-level="11.5.3" data-path="casp8-5.html"><a href="casp8-5.html#genomic-binding-sites-4"><i class="fa fa-check"></i><b>11.5.3</b> Genomic Binding Sites</a></li>
<li class="chapter" data-level="11.5.4" data-path="casp8-5.html"><a href="casp8-5.html#z-score-4"><i class="fa fa-check"></i><b>11.5.4</b> Z-Score</a></li>
<li class="chapter" data-level="11.5.5" data-path="casp8-5.html"><a href="casp8-5.html#efficiency-scores-4"><i class="fa fa-check"></i><b>11.5.5</b> Efficiency Scores</a></li>
<li class="chapter" data-level="11.5.6" data-path="casp8-5.html"><a href="casp8-5.html#e-crisp-scores-4"><i class="fa fa-check"></i><b>11.5.6</b> E-CRISP Scores</a></li>
<li class="chapter" data-level="11.5.7" data-path="casp8-5.html"><a href="casp8-5.html#sgrna-sequences-4"><i class="fa fa-check"></i><b>11.5.7</b> sgRNA Sequences</a></li>
<li class="chapter" data-level="11.5.8" data-path="casp8-5.html"><a href="casp8-5.html#sgrna-binding-sites-4"><i class="fa fa-check"></i><b>11.5.8</b> sgRNA Binding Sites</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="gene-set-enrichment-analysis.html"><a href="gene-set-enrichment-analysis.html"><i class="fa fa-check"></i><b>12</b> Gene Set Enrichment Analysis</a><ul>
<li class="chapter" data-level="12.1" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html"><i class="fa fa-check"></i><b>12.1</b> TNFRSF10A, TNFRSF10B, TNFRSF10C (1)</a><ul>
<li class="chapter" data-level="12.1.1" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html#pathways"><i class="fa fa-check"></i><b>12.1.1</b> Pathways</a></li>
<li class="chapter" data-level="12.1.2" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html#transcription"><i class="fa fa-check"></i><b>12.1.2</b> Transcription</a></li>
<li class="chapter" data-level="12.1.3" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html#ontologies"><i class="fa fa-check"></i><b>12.1.3</b> Ontologies</a></li>
<li class="chapter" data-level="12.1.4" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html#diseases"><i class="fa fa-check"></i><b>12.1.4</b> Diseases</a></li>
<li class="chapter" data-level="12.1.5" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-1.html#cell-types"><i class="fa fa-check"></i><b>12.1.5</b> Cell Types</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="gene-comparisons.html"><a href="gene-comparisons.html"><i class="fa fa-check"></i><b>13</b> Gene Comparisons</a><ul>
<li class="chapter" data-level="13.1" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html"><i class="fa fa-check"></i><b>13.1</b> TNFRSF10A, TNFRSF10B, TNFRSF10C, CASP3, BAX, BCL2L1, MCL1 (1)</a><ul>
<li class="chapter" data-level="13.1.1" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html#readcount-untreated-group"><i class="fa fa-check"></i><b>13.1.1</b> Readcount Untreated Group</a></li>
<li class="chapter" data-level="13.1.2" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html#readcount-treated-group"><i class="fa fa-check"></i><b>13.1.2</b> Readcount Treated Group</a></li>
<li class="chapter" data-level="13.1.3" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html#log2-foldchange-5"><i class="fa fa-check"></i><b>13.1.3</b> Log2 Foldchange</a></li>
<li class="chapter" data-level="13.1.4" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html#z-scores"><i class="fa fa-check"></i><b>13.1.4</b> Z-Scores</a></li>
<li class="chapter" data-level="13.1.5" data-path="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html"><a href="tnfrsf10a-tnfrsf10b-tnfrsf10c-casp3-bax-bcl2l1-mcl1-1.html#sgrna-binding-sites-5"><i class="fa fa-check"></i><b>13.1.5</b> sgRNA Binding Sites</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="credits.html"><a href="credits.html"><i class="fa fa-check"></i><b>14</b> Credits</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">CRISPR-AnalyzeR Report</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="overview-1" class="section level2">
<h2><span class="header-section-number">6.1</span> Overview</h2>
<p>Normalized readcount foldchanges or Z-Ratios between all samples on both gene and sgRNA level are shown in this section.</p>
<div id="foldchange-of-gene-effects" class="section level3">
<h3><span class="header-section-number">6.1.1</span> Foldchange of Gene Effects</h3>
<div id="htmlwidget-f696de82f7ae5ec325ca" style="width:100%;height:500px;" class="highchart html-widget"></div>
<script type="application/json" data-for="htmlwidget-f696de82f7ae5ec325ca">{"x":{"hc_opts":{"title":{"text":"log2fc Distribution"},"yAxis":{"title":{"text":null}},"credits":{"enabled":false},"exporting":{"enabled":true,"printMaxWidth":2000,"scale":8,"formAttributes":{"target":"_blank"}},"plotOptions":{"series":{"turboThreshold":0},"treemap":{"layoutAlgorithm":"squarified"},"bubble":{"minSize":5,"maxSize":25},"boxplot":{"fillColor":"white","lineWidth":1}},"annotationsOptions":{"enabledButtons":false},"tooltip":{"delayForDisplay":10},"chart":{"type":"boxplot","zoomType":"x"},"xAxis":{"categories":"value","title":{"text":"Dataset"}},"series":[{"data":[[-0.31,-0.08,0.01,0.08,0.31]],"type":"boxplot","name":"PBS_Replicate1 vs. PBS_Replicate2"},{"data":[[-0.87,-0.23,-0.01,0.22,0.84]],"type":"boxplot","name":"PBS_Replicate1 vs. TRAIL_Replicate2"},{"data":[[-1.04,-0.27,-0.01,0.26,1.03]],"type":"boxplot","name":"PBS_Replicate1 vs. TRAIL_Replicate1"},{"data":[[-0.85,-0.21,-0,0.22,0.83]],"type":"boxplot","name":"PBS_Replicate2 vs. TRAIL_Replicate2"},{"data":[[-0.93,-0.25,0,0.23,0.94]],"type":"boxplot","name":"PBS_Replicate2 vs. TRAIL_Replicate1"},{"data":[[-0.59,-0.16,-0.02,0.13,0.53]],"type":"boxplot","name":"TRAIL_Replicate2 vs. TRAIL_Replicate1"}],"legend":{"enabled":["function (x, y = NULL, legend, fill = NULL, col = par(\"col\"), ","    border = \"black\", lty, lwd, pch, angle = 45, density = NULL, ","    bty = \"o\", bg = par(\"bg\"), box.lwd = par(\"lwd\"), box.lty = par(\"lty\"), ","    box.col = par(\"fg\"), pt.bg = NA, cex = 1, pt.cex = cex, pt.lwd = lwd, ","    xjust = 0, yjust = 1, x.intersp = 1, y.intersp = 1, adj = c(0, ","        0.5), text.width = NULL, text.col = par(\"col\"), text.font = NULL, ","    merge = do.lines && has.pch, trace = FALSE, plot = TRUE, ","    ncol = 1, horiz = FALSE, title = NULL, inset = 0, xpd, title.col = text.col, ","    title.adj = 0.5, seg.len = 2) ","{","    if (missing(legend) && !missing(y) && (is.character(y) || ","        is.expression(y))) {","        legend <- y","        y <- NULL","    }","    mfill <- !missing(fill) || !missing(density)","    if (!missing(xpd)) {","        op <- par(\"xpd\")","        on.exit(par(xpd = op))","        par(xpd = xpd)","    }","    title <- as.graphicsAnnot(title)","    if (length(title) > 1) ","        stop(\"invalid 'title'\")","    legend <- as.graphicsAnnot(legend)","    n.leg <- if (is.call(legend)) ","        1","    else length(legend)","    if (n.leg == 0) ","        stop(\"'legend' is of length 0\")","    auto <- if (is.character(x)) ","        match.arg(x, c(\"bottomright\", \"bottom\", \"bottomleft\", ","            \"left\", \"topleft\", \"top\", \"topright\", \"right\", \"center\"))","    else NA","    if (is.na(auto)) {","        xy <- xy.coords(x, y)","        x <- xy$x","        y <- xy$y","        nx <- length(x)","        if (nx < 1 || nx > 2) ","            stop(\"invalid coordinate lengths\")","    }","    else nx <- 0","    xlog <- par(\"xlog\")","    ylog <- par(\"ylog\")","    rect2 <- function(left, top, dx, dy, density = NULL, angle, ","        ...) {","        r <- left + dx","        if (xlog) {","            left <- 10^left","            r <- 10^r","        }","        b <- top - dy","        if (ylog) {","            top <- 10^top","            b <- 10^b","        }","        rect(left, top, r, b, angle = angle, density = density, ","            ...)","    }","    segments2 <- function(x1, y1, dx, dy, ...) {","        x2 <- x1 + dx","        if (xlog) {","            x1 <- 10^x1","            x2 <- 10^x2","        }","        y2 <- y1 + dy","        if (ylog) {","            y1 <- 10^y1","            y2 <- 10^y2","        }","        segments(x1, y1, x2, y2, ...)","    }","    points2 <- function(x, y, ...) {","        if (xlog) ","            x <- 10^x","        if (ylog) ","            y <- 10^y","        points(x, y, ...)","    }","    text2 <- function(x, y, ...) {","        if (xlog) ","            x <- 10^x","        if (ylog) ","            y <- 10^y","        text(x, y, ...)","    }","    if (trace) ","        catn <- function(...) do.call(\"cat\", c(lapply(list(...), ","            formatC), list(\"\\n\")))","    cin <- par(\"cin\")","    Cex <- cex * par(\"cex\")","    if (is.null(text.width)) ","        text.width <- max(abs(strwidth(legend, units = \"user\", ","            cex = cex, font = text.font)))","    else if (!is.numeric(text.width) || text.width < 0) ","        stop(\"'text.width' must be numeric, >= 0\")","    xc <- Cex * xinch(cin[1L], warn.log = FALSE)","    yc <- Cex * yinch(cin[2L], warn.log = FALSE)","    if (xc < 0) ","        text.width <- -text.width","    xchar <- xc","    xextra <- 0","    yextra <- yc * (y.intersp - 1)","    ymax <- yc * max(1, strheight(legend, units = \"user\", cex = cex)/yc)","    ychar <- yextra + ymax","    if (trace) ","        catn(\"  xchar=\", xchar, \"; (yextra,ychar)=\", c(yextra, ","            ychar))","    if (mfill) {","        xbox <- xc * 0.8","        ybox <- yc * 0.5","        dx.fill <- xbox","    }","    do.lines <- (!missing(lty) && (is.character(lty) || any(lty > ","        0))) || !missing(lwd)","    n.legpercol <- if (horiz) {","        if (ncol != 1) ","            warning(gettextf(\"horizontal specification overrides: Number of columns := %d\", ","                n.leg), domain = NA)","        ncol <- n.leg","        1","    }","    else ceiling(n.leg/ncol)","    has.pch <- !missing(pch) && length(pch) > 0","    if (do.lines) {","        x.off <- if (merge) ","            -0.7","        else 0","    }","    else if (merge) ","        warning(\"'merge = TRUE' has no effect when no line segments are drawn\")","    if (has.pch) {","        if (is.character(pch) && !is.na(pch[1L]) && nchar(pch[1L], ","            type = \"c\") > 1) {","            if (length(pch) > 1) ","                warning(\"not using pch[2..] since pch[1L] has multiple chars\")","            np <- nchar(pch[1L], type = \"c\")","            pch <- substr(rep.int(pch[1L], np), 1L:np, 1L:np)","        }","        if (!is.character(pch)) ","            pch <- as.integer(pch)","    }","    if (is.na(auto)) {","        if (xlog) ","            x <- log10(x)","        if (ylog) ","            y <- log10(y)","    }","    if (nx == 2) {","        x <- sort(x)","        y <- sort(y)","        left <- x[1L]","        top <- y[2L]","        w <- diff(x)","        h <- diff(y)","        w0 <- w/ncol","        x <- mean(x)","        y <- mean(y)","        if (missing(xjust)) ","            xjust <- 0.5","        if (missing(yjust)) ","            yjust <- 0.5","    }","    else {","        h <- (n.legpercol + (!is.null(title))) * ychar + yc","        w0 <- text.width + (x.intersp + 1) * xchar","        if (mfill) ","            w0 <- w0 + dx.fill","        if (do.lines) ","            w0 <- w0 + (seg.len + x.off) * xchar","        w <- ncol * w0 + 0.5 * xchar","        if (!is.null(title) && (abs(tw <- strwidth(title, units = \"user\", ","            cex = cex) + 0.5 * xchar)) > abs(w)) {","            xextra <- (tw - w)/2","            w <- tw","        }","        if (is.na(auto)) {","            left <- x - xjust * w","            top <- y + (1 - yjust) * h","        }","        else {","            usr <- par(\"usr\")","            inset <- rep_len(inset, 2)","            insetx <- inset[1L] * (usr[2L] - usr[1L])","            left <- switch(auto, bottomright = , topright = , ","                right = usr[2L] - w - insetx, bottomleft = , ","                left = , topleft = usr[1L] + insetx, bottom = , ","                top = , center = (usr[1L] + usr[2L] - w)/2)","            insety <- inset[2L] * (usr[4L] - usr[3L])","            top <- switch(auto, bottomright = , bottom = , bottomleft = usr[3L] + ","                h + insety, topleft = , top = , topright = usr[4L] - ","                insety, left = , right = , center = (usr[3L] + ","                usr[4L] + h)/2)","        }","    }","    if (plot && bty != \"n\") {","        if (trace) ","            catn(\"  rect2(\", left, \",\", top, \", w=\", w, \", h=\", ","                h, \", ...)\", sep = \"\")","        rect2(left, top, dx = w, dy = h, col = bg, density = NULL, ","            lwd = box.lwd, lty = box.lty, border = box.col)","    }","    xt <- left + xchar + xextra + (w0 * rep.int(0:(ncol - 1), ","        rep.int(n.legpercol, ncol)))[1L:n.leg]","    yt <- top - 0.5 * yextra - ymax - (rep.int(1L:n.legpercol, ","        ncol)[1L:n.leg] - 1 + (!is.null(title))) * ychar","    if (mfill) {","        if (plot) {","            if (!is.null(fill)) ","                fill <- rep_len(fill, n.leg)","            rect2(left = xt, top = yt + ybox/2, dx = xbox, dy = ybox, ","                col = fill, density = density, angle = angle, ","                border = border)","        }","        xt <- xt + dx.fill","    }","    if (plot && (has.pch || do.lines)) ","        col <- rep_len(col, n.leg)","    if (missing(lwd) || is.null(lwd)) ","        lwd <- par(\"lwd\")","    if (do.lines) {","        if (missing(lty) || is.null(lty)) ","            lty <- 1","        lty <- rep_len(lty, n.leg)","        lwd <- rep_len(lwd, n.leg)","        ok.l <- !is.na(lty) & (is.character(lty) | lty > 0) & ","            !is.na(lwd)","        if (trace) ","            catn(\"  segments2(\", xt[ok.l] + x.off * xchar, \",\", ","                yt[ok.l], \", dx=\", seg.len * xchar, \", dy=0, ...)\")","        if (plot) ","            segments2(xt[ok.l] + x.off * xchar, yt[ok.l], dx = seg.len * ","                xchar, dy = 0, lty = lty[ok.l], lwd = lwd[ok.l], ","                col = col[ok.l])","        xt <- xt + (seg.len + x.off) * xchar","    }","    if (has.pch) {","        pch <- rep_len(pch, n.leg)","        pt.bg <- rep_len(pt.bg, n.leg)","        pt.cex <- rep_len(pt.cex, n.leg)","        pt.lwd <- rep_len(pt.lwd, n.leg)","        ok <- !is.na(pch)","        if (!is.character(pch)) {","            ok <- ok & (pch >= 0 | pch <= -32)","        }","        else {","            ok <- ok & nzchar(pch)","        }","        x1 <- (if (merge && do.lines) ","            xt - (seg.len/2) * xchar","        else xt)[ok]","        y1 <- yt[ok]","        if (trace) ","            catn(\"  points2(\", x1, \",\", y1, \", pch=\", pch[ok], ","                \", ...)\")","        if (plot) ","            points2(x1, y1, pch = pch[ok], col = col[ok], cex = pt.cex[ok], ","                bg = pt.bg[ok], lwd = pt.lwd[ok])","    }","    xt <- xt + x.intersp * xchar","    if (plot) {","        if (!is.null(title)) ","            text2(left + w * title.adj, top - ymax, labels = title, ","                adj = c(title.adj, 0), cex = cex, col = title.col)","        text2(xt, yt, labels = legend, adj = adj, cex = cex, ","            col = text.col, font = text.font)","    }","    invisible(list(rect = list(w = w, h = h, left = left, top = top), ","        text = list(x = xt, y = yt)))","}"]}},"theme":{"colors":["#F44336","#2196F3","#4CAF50","#FFEB3B","#9C27B0","#03A9F4","#8BC34A","#FF9800","#3F51B5","#E91E63","#CDDC39","#FF5722"],"chart":{"style":{"fontFamily":"Roboto","color":"#212121"}},"xAxis":{"gridLineWidth":1,"gridLineColor":"#F3F3F3","lineColor":"#F3F3F3","minorGridLineColor":"#F3F3F3","tickColor":"#F3F3F3","tickWidth":1},"yAxis":{"gridLineColor":"#F3F3F3","lineColor":"#F3F3F3","minorGridLineColor":"#F3F3F3","tickColor":"#F3F3F3","tickWidth":1},"plotOptions":{"boxplot":{"fillColor":"#FAFAFA"}},"legendBackgroundColor":"rgba(0, 0, 0, 0.5)","background2":"#505053","dataLabelsColor":"#B0B0B3","textColor":"#C0C0C0","contrastTextColor":"#F0F0F3","maskColor":"rgba(255,255,255,0.3)"},"conf_opts":{"global":{"Date":null,"VMLRadialGradientURL":"http =//code.highcharts.com/list(version)/gfx/vml-radial-gradient.png","canvasToolsURL":"http =//code.highcharts.com/list(version)/modules/canvas-tools.js","getTimezoneOffset":null,"timezoneOffset":0,"useUTC":true},"lang":{"contextButtonTitle":"Chart context menu","decimalPoint":".","downloadJPEG":"Download JPEG image","downloadPDF":"Download PDF document","downloadPNG":"Download PNG image","downloadSVG":"Download SVG vector image","drillUpText":"Back to {series.name}","invalidDate":null,"loading":"Loading...","months":["January","February","March","April","May","June","July","August","September","October","November","December"],"noData":"No data to display","numericSymbols":["k","M","G","T","P","E"],"printChart":"Print chart","resetZoom":"Reset zoom","resetZoomTitle":"Reset zoom level 1:1","shortMonths":["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"thousandsSep":" ","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"]}},"type":"chart","fonts":"Roboto","debug":false},"evals":[],"jsHooks":[]}</script>
</div>
<div id="z-ratio-of-gene-effects" class="section level3">
<h3><span class="header-section-number">6.1.2</span> Z-Ratio of Gene Effects</h3>
<div id="htmlwidget-f8de3ded7657ecb100b8" style="width:100%;height:500px;" class="highchart html-widget"></div>
<script type="application/json" data-for="htmlwidget-f8de3ded7657ecb100b8">{"x":{"hc_opts":{"title":{"text":"zratio Distribution"},"yAxis":{"title":{"text":null}},"credits":{"enabled":false},"exporting":{"enabled":true,"printMaxWidth":2000,"scale":8,"formAttributes":{"target":"_blank"}},"plotOptions":{"series":{"turboThreshold":0},"treemap":{"layoutAlgorithm":"squarified"},"bubble":{"minSize":5,"maxSize":25},"boxplot":{"fillColor":"white","lineWidth":1}},"annotationsOptions":{"enabledButtons":false},"tooltip":{"delayForDisplay":10},"chart":{"type":"boxplot","zoomType":"x"},"xAxis":{"categories":"value","title":{"text":"Dataset"}},"series":[{"data":[[-2.19555343808927,-0.551966183979648,0.0156765784730842,0.558446498227325,1.96742120386589]],"type":"boxplot","name":"PBS_Replicate1 vs. PBS_Replicate2"},{"data":[[-2.02311876171345,-0.516704341398236,0.0355993230181696,0.518445291642724,1.92543229953734]],"type":"boxplot","name":"PBS_Replicate1 vs. TRAIL_Replicate2"},{"data":[[-2.24062777823325,-0.539583441601058,0.0520103549243315,0.625621796478825,2.20764955260862]],"type":"boxplot","name":"PBS_Replicate1 vs. TRAIL_Replicate1"},{"data":[[-2.0568750666625,-0.529360397965148,0.0483926917013587,0.54154197164615,2.14337548084782]],"type":"boxplot","name":"PBS_Replicate2 vs. TRAIL_Replicate2"},{"data":[[-2.25898842867486,-0.551673584162923,-0.00455795367880421,0.587463436773542,2.1494380936652]],"type":"boxplot","name":"PBS_Replicate2 vs. TRAIL_Replicate1"},{"data":[[-2.26727745701738,-0.645277232785576,-0.00791035511157861,0.615856654758863,2.32397904910409]],"type":"boxplot","name":"TRAIL_Replicate2 vs. TRAIL_Replicate1"}],"legend":{"enabled":["function (x, y = NULL, legend, fill = NULL, col = par(\"col\"), ","    border = \"black\", lty, lwd, pch, angle = 45, density = NULL, ","    bty = \"o\", bg = par(\"bg\"), box.lwd = par(\"lwd\"), box.lty = par(\"lty\"), ","    box.col = par(\"fg\"), pt.bg = NA, cex = 1, pt.cex = cex, pt.lwd = lwd, ","    xjust = 0, yjust = 1, x.intersp = 1, y.intersp = 1, adj = c(0, ","        0.5), text.width = NULL, text.col = par(\"col\"), text.font = NULL, ","    merge = do.lines && has.pch, trace = FALSE, plot = TRUE, ","    ncol = 1, horiz = FALSE, title = NULL, inset = 0, xpd, title.col = text.col, ","    title.adj = 0.5, seg.len = 2) ","{","    if (missing(legend) && !missing(y) && (is.character(y) || ","        is.expression(y))) {","        legend <- y","        y <- NULL","    }","    mfill <- !missing(fill) || !missing(density)","    if (!missing(xpd)) {","        op <- par(\"xpd\")","        on.exit(par(xpd = op))","        par(xpd = xpd)","    }","    title <- as.graphicsAnnot(title)","    if (length(title) > 1) ","        stop(\"invalid 'title'\")","    legend <- as.graphicsAnnot(legend)","    n.leg <- if (is.call(legend)) ","        1","    else length(legend)","    if (n.leg == 0) ","        stop(\"'legend' is of length 0\")","    auto <- if (is.character(x)) ","        match.arg(x, c(\"bottomright\", \"bottom\", \"bottomleft\", ","            \"left\", \"topleft\", \"top\", \"topright\", \"right\", \"center\"))","    else NA","    if (is.na(auto)) {","        xy <- xy.coords(x, y)","        x <- xy$x","        y <- xy$y","        nx <- length(x)","        if (nx < 1 || nx > 2) ","            stop(\"invalid coordinate lengths\")","    }","    else nx <- 0","    xlog <- par(\"xlog\")","    ylog <- par(\"ylog\")","    rect2 <- function(left, top, dx, dy, density = NULL, angle, ","        ...) {","        r <- left + dx","        if (xlog) {","            left <- 10^left","            r <- 10^r","        }","        b <- top - dy","        if (ylog) {","            top <- 10^top","            b <- 10^b","        }","        rect(left, top, r, b, angle = angle, density = density, ","            ...)","    }","    segments2 <- function(x1, y1, dx, dy, ...) {","        x2 <- x1 + dx","        if (xlog) {","            x1 <- 10^x1","            x2 <- 10^x2","        }","        y2 <- y1 + dy","        if (ylog) {","            y1 <- 10^y1","            y2 <- 10^y2","        }","        segments(x1, y1, x2, y2, ...)","    }","    points2 <- function(x, y, ...) {","        if (xlog) ","            x <- 10^x","        if (ylog) ","            y <- 10^y","        points(x, y, ...)","    }","    text2 <- function(x, y, ...) {","        if (xlog) ","            x <- 10^x","        if (ylog) ","            y <- 10^y","        text(x, y, ...)","    }","    if (trace) ","        catn <- function(...) do.call(\"cat\", c(lapply(list(...), ","            formatC), list(\"\\n\")))","    cin <- par(\"cin\")","    Cex <- cex * par(\"cex\")","    if (is.null(text.width)) ","        text.width <- max(abs(strwidth(legend, units = \"user\", ","            cex = cex, font = text.font)))","    else if (!is.numeric(text.width) || text.width < 0) ","        stop(\"'text.width' must be numeric, >= 0\")","    xc <- Cex * xinch(cin[1L], warn.log = FALSE)","    yc <- Cex * yinch(cin[2L], warn.log = FALSE)","    if (xc < 0) ","        text.width <- -text.width","    xchar <- xc","    xextra <- 0","    yextra <- yc * (y.intersp - 1)","    ymax <- yc * max(1, strheight(legend, units = \"user\", cex = cex)/yc)","    ychar <- yextra + ymax","    if (trace) ","        catn(\"  xchar=\", xchar, \"; (yextra,ychar)=\", c(yextra, ","            ychar))","    if (mfill) {","        xbox <- xc * 0.8","        ybox <- yc * 0.5","        dx.fill <- xbox","    }","    do.lines <- (!missing(lty) && (is.character(lty) || any(lty > ","        0))) || !missing(lwd)","    n.legpercol <- if (horiz) {","        if (ncol != 1) ","            warning(gettextf(\"horizontal specification overrides: Number of columns := %d\", ","                n.leg), domain = NA)","        ncol <- n.leg","        1","    }","    else ceiling(n.leg/ncol)","    has.pch <- !missing(pch) && length(pch) > 0","    if (do.lines) {","        x.off <- if (merge) ","            -0.7","        else 0","    }","    else if (merge) ","        warning(\"'merge = TRUE' has no effect when no line segments are drawn\")","    if (has.pch) {","        if (is.character(pch) && !is.na(pch[1L]) && nchar(pch[1L], ","            type = \"c\") > 1) {","            if (length(pch) > 1) ","                warning(\"not using pch[2..] since pch[1L] has multiple chars\")","            np <- nchar(pch[1L], type = \"c\")","            pch <- substr(rep.int(pch[1L], np), 1L:np, 1L:np)","        }","        if (!is.character(pch)) ","            pch <- as.integer(pch)","    }","    if (is.na(auto)) {","        if (xlog) ","            x <- log10(x)","        if (ylog) ","            y <- log10(y)","    }","    if (nx == 2) {","        x <- sort(x)","        y <- sort(y)","        left <- x[1L]","        top <- y[2L]","        w <- diff(x)","        h <- diff(y)","        w0 <- w/ncol","        x <- mean(x)","        y <- mean(y)","        if (missing(xjust)) ","            xjust <- 0.5","        if (missing(yjust)) ","            yjust <- 0.5","    }","    else {","        h <- (n.legpercol + (!is.null(title))) * ychar + yc","        w0 <- text.width + (x.intersp + 1) * xchar","        if (mfill) ","            w0 <- w0 + dx.fill","        if (do.lines) ","            w0 <- w0 + (seg.len + x.off) * xchar","        w <- ncol * w0 + 0.5 * xchar","        if (!is.null(title) && (abs(tw <- strwidth(title, units = \"user\", ","            cex = cex) + 0.5 * xchar)) > abs(w)) {","            xextra <- (tw - w)/2","            w <- tw","        }","        if (is.na(auto)) {","            left <- x - xjust * w","            top <- y + (1 - yjust) * h","        }","        else {","            usr <- par(\"usr\")","            inset <- rep_len(inset, 2)","            insetx <- inset[1L] * (usr[2L] - usr[1L])","            left <- switch(auto, bottomright = , topright = , ","                right = usr[2L] - w - insetx, bottomleft = , ","                left = , topleft = usr[1L] + insetx, bottom = , ","                top = , center = (usr[1L] + usr[2L] - w)/2)","            insety <- inset[2L] * (usr[4L] - usr[3L])","            top <- switch(auto, bottomright = , bottom = , bottomleft = usr[3L] + ","                h + insety, topleft = , top = , topright = usr[4L] - ","                insety, left = , right = , center = (usr[3L] + ","                usr[4L] + h)/2)","        }","    }","    if (plot && bty != \"n\") {","        if (trace) ","            catn(\"  rect2(\", left, \",\", top, \", w=\", w, \", h=\", ","                h, \", ...)\", sep = \"\")","        rect2(left, top, dx = w, dy = h, col = bg, density = NULL, ","            lwd = box.lwd, lty = box.lty, border = box.col)","    }","    xt <- left + xchar + xextra + (w0 * rep.int(0:(ncol - 1), ","        rep.int(n.legpercol, ncol)))[1L:n.leg]","    yt <- top - 0.5 * yextra - ymax - (rep.int(1L:n.legpercol, ","        ncol)[1L:n.leg] - 1 + (!is.null(title))) * ychar","    if (mfill) {","        if (plot) {","            if (!is.null(fill)) ","                fill <- rep_len(fill, n.leg)","            rect2(left = xt, top = yt + ybox/2, dx = xbox, dy = ybox, ","                col = fill, density = density, angle = angle, ","                border = border)","        }","        xt <- xt + dx.fill","    }","    if (plot && (has.pch || do.lines)) ","        col <- rep_len(col, n.leg)","    if (missing(lwd) || is.null(lwd)) ","        lwd <- par(\"lwd\")","    if (do.lines) {","        if (missing(lty) || is.null(lty)) ","            lty <- 1","        lty <- rep_len(lty, n.leg)","        lwd <- rep_len(lwd, n.leg)","        ok.l <- !is.na(lty) & (is.character(lty) | lty > 0) & ","            !is.na(lwd)","        if (trace) ","            catn(\"  segments2(\", xt[ok.l] + x.off * xchar, \",\", ","                yt[ok.l], \", dx=\", seg.len * xchar, \", dy=0, ...)\")","        if (plot) ","            segments2(xt[ok.l] + x.off * xchar, yt[ok.l], dx = seg.len * ","                xchar, dy = 0, lty = lty[ok.l], lwd = lwd[ok.l], ","                col = col[ok.l])","        xt <- xt + (seg.len + x.off) * xchar","    }","    if (has.pch) {","        pch <- rep_len(pch, n.leg)","        pt.bg <- rep_len(pt.bg, n.leg)","        pt.cex <- rep_len(pt.cex, n.leg)","        pt.lwd <- rep_len(pt.lwd, n.leg)","        ok <- !is.na(pch)","        if (!is.character(pch)) {","            ok <- ok & (pch >= 0 | pch <= -32)","        }","        else {","            ok <- ok & nzchar(pch)","        }","        x1 <- (if (merge && do.lines) ","            xt - (seg.len/2) * xchar","        else xt)[ok]","        y1 <- yt[ok]","        if (trace) ","            catn(\"  points2(\", x1, \",\", y1, \", pch=\", pch[ok], ","                \", ...)\")","        if (plot) ","            points2(x1, y1, pch = pch[ok], col = col[ok], cex = pt.cex[ok], ","                bg = pt.bg[ok], lwd = pt.lwd[ok])","    }","    xt <- xt + x.intersp * xchar","    if (plot) {","        if (!is.null(title)) ","            text2(left + w * title.adj, top - ymax, labels = title, ","                adj = c(title.adj, 0), cex = cex, col = title.col)","        text2(xt, yt, labels = legend, adj = adj, cex = cex, ","            col = text.col, font = text.font)","    }","    invisible(list(rect = list(w = w, h = h, left = left, top = top), ","        text = list(x = xt, y = yt)))","}"]}},"theme":{"colors":["#F44336","#2196F3","#4CAF50","#FFEB3B","#9C27B0","#03A9F4","#8BC34A","#FF9800","#3F51B5","#E91E63","#CDDC39","#FF5722"],"chart":{"style":{"fontFamily":"Roboto","color":"#212121"}},"xAxis":{"gridLineWidth":1,"gridLineColor":"#F3F3F3","lineColor":"#F3F3F3","minorGridLineColor":"#F3F3F3","tickColor":"#F3F3F3","tickWidth":1},"yAxis":{"gridLineColor":"#F3F3F3","lineColor":"#F3F3F3","minorGridLineColor":"#F3F3F3","tickColor":"#F3F3F3","tickWidth":1},"plotOptions":{"boxplot":{"fillColor":"#FAFAFA"}},"legendBackgroundColor":"rgba(0, 0, 0, 0.5)","background2":"#505053","dataLabelsColor":"#B0B0B3","textColor":"#C0C0C0","contrastTextColor":"#F0F0F3","maskColor":"rgba(255,255,255,0.3)"},"conf_opts":{"global":{"Date":null,"VMLRadialGradientURL":"http =//code.highcharts.com/list(version)/gfx/vml-radial-gradient.png","canvasToolsURL":"http =//code.highcharts.com/list(version)/modules/canvas-tools.js","getTimezoneOffset":null,"timezoneOffset":0,"useUTC":true},"lang":{"contextButtonTitle":"Chart context menu","decimalPoint":".","downloadJPEG":"Download JPEG image","downloadPDF":"Download PDF document","downloadPNG":"Download PNG image","downloadSVG":"Download SVG vector image","drillUpText":"Back to {series.name}","invalidDate":null,"loading":"Loading...","months":["January","February","March","April","May","June","July","August","September","October","November","December"],"noData":"No data to display","numericSymbols":["k","M","G","T","P","E"],"printChart":"Print chart","resetZoom":"Reset zoom","resetZoomTitle":"Reset zoom level 1:1","shortMonths":["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"thousandsSep":" ","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"]}},"type":"chart","fonts":"Roboto","debug":false},"evals":[],"jsHooks":[]}</script>
</div>
<div id="foldchange-of-sgrna-effects" class="section level3">
<h3><span class="header-section-number">6.1.3</span> Foldchange of sgRNA Effects</h3>
<div id="htmlwidget-aa7a2c3b3c45df381adb" style="width:100%;height:500px;" class="highchart html-widget"></div>
<script type="application/json" data-for="htmlwidget-aa7a2c3b3c45df381adb">{"x":{"hc_opts":{"title":{"text":"log2fc Distribution"},"yAxis":{"title":{"text":null}},"credits":{"enabled":false},"exporting":{"enabled":true,"printMaxWidth":2000,"scale":8,"formAttributes":{"target":"_blank"}},"plotOptions":{"series":{"turboThreshold":0},"treemap":{"layoutAlgorithm":"squarified"},"bubble":{"minSize":5,"maxSize":25},"boxplot":{"fillColor":"white","lineWidth":1}},"annotationsOptions":{"enabledButtons":false},"tooltip":{"delayForDisplay":10},"chart":{"type":"boxplot","zoomType":"x"},"xAxis":{"categories":"value","title":{"text":"Dataset"}},"series":[{"data":[[-1.67,-0.41,0.01,0.43,1.69]],"type":"boxplot","name":"PBS_Replicate1 vs. PBS_Replicate2"},{"data":[[-3.82,-0.89,0.04,1.07,4.01]],"type":"boxplot","name":"PBS_Replicate1 vs. TRAIL_Replicate2"},{"data":[[-4.34,-1,0.05,1.23,4.57]],"type":"boxplot","name":"PBS_Replicate1 vs. TRAIL_Replicate1"},{"data":[[-3.84,-0.89,0.03,1.08,4.02]],"type":"boxplot","name":"PBS_Replicate2 vs. TRAIL_Replicate2"},{"data":[[-4.28,-1,0.03,1.19,4.47]],"type":"boxplot","name":"PBS_Replicate2 vs. TRAIL_Replicate1"},{"data":[[-3.05,-0.76,-0.02,0.77,3.06]],"type":"boxplot","name":"TRAIL_Replicate2 vs. TRAIL_Replicate1"}],"legend":{"enabled":["function (x, y = NULL, legend, fill = NULL, col = par(\"col\"), ","    border = \"black\", lty, lwd, pch, angle = 45, density = NULL, ","    bty = \"o\", bg = par(\"bg\"), box.lwd = par(\"lwd\"), box.lty = par(\"lty\"), ","    box.col = par(\"fg\"), pt.bg = NA, cex = 1, pt.cex = cex, pt.lwd = lwd, ","    xjust = 0, yjust = 1, x.intersp = 1, y.intersp = 1, adj = c(0, ","        0.5), text.width = NULL, text.col = par(\"col\"), text.font = NULL, ","    merge = do.lines && has.pch, trace = FALSE, plot = TRUE, ","    ncol = 1, horiz = FALSE, title = NULL, inset = 0, xpd, title.col = text.col, ","    title.adj = 0.5, seg.len = 2) ","{","    if (missing(legend) && !missing(y) && (is.character(y) || ","        is.expression(y))) {","        legend <- y","        y <- NULL","    }","    mfill <- !missing(fill) || !missing(density)","    if (!missing(xpd)) {","        op <- par(\"xpd\")","        on.exit(par(xpd = op))","        par(xpd = xpd)","    }","    title <- as.graphicsAnnot(title)","    if (length(title) > 1) ","        stop(\"invalid 'title'\")","    legend <- as.graphicsAnnot(legend)","    n.leg <- if (is.call(legend)) ","        1","    else length(legend)","    if (n.leg == 0) ","        stop(\"'legend' is of length 0\")","    auto <- if (is.character(x)) ","        match.arg(x, c(\"bottomright\", \"bottom\", \"bottomleft\", ","            \"left\", \"topleft\", \"top\", \"topright\", \"right\", \"center\"))","    else NA","    if (is.na(auto)) {","        xy <- xy.coords(x, y)","        x <- xy$x","        y <- xy$y","        nx <- length(x)","        if (nx < 1 || nx > 2) ","            stop(\"invalid coordinate lengths\")","    }","    else nx <- 0","    xlog <- par(\"xlog\")","    ylog <- par(\"ylog\")","    rect2 <- function(left, top, dx, dy, density = NULL, angle, ","        ...) {","        r <- left + dx","        if (xlog) {","            left <- 10^left","            r <- 10^r","        }","        b <- top - dy","        if (ylog) {","            top <- 10^top","            b <- 10^b","        }","        rect(left, top, r, b, angle = angle, density = density, ","            ...)","    }","    segments2 <- function(x1, y1, dx, dy, ...) {","        x2 <- x1 + dx","        if (xlog) {","            x1 <- 10^x1","            x2 <- 10^x2","        }","        y2 <- y1 + dy","        if (ylog) {","            y1 <- 10^y1","            y2 <- 10^y2","        }","        segments(x1, y1, x2, y2, ...)","    }","    points2 <- function(x, y, ...) {","        if (xlog) ","            x <- 10^x","        if (ylog) ","            y <- 10^y","        points(x, y, ...)","    }","    text2 <- function(x, y, ...) {","        if (xlog) ","            x <- 10^x","        if (ylog) ","            y <- 10^y","        text(x, y, ...)","    }","    if (trace) ","        catn <- function(...) do.call(\"cat\", c(lapply(list(...), ","            formatC), list(\"\\n\")))","    cin <- par(\"cin\")","    Cex <- cex * par(\"cex\")","    if (is.null(text.width)) ","        text.width <- max(abs(strwidth(legend, units = \"user\", ","            cex = cex, font = text.font)))","    else if (!is.numeric(text.width) || text.width < 0) ","        stop(\"'text.width' must be numeric, >= 0\")","    xc <- Cex * xinch(cin[1L], warn.log = FALSE)","    yc <- Cex * yinch(cin[2L], warn.log = FALSE)","    if (xc < 0) ","        text.width <- -text.width","    xchar <- xc","    xextra <- 0","    yextra <- yc * (y.intersp - 1)","    ymax <- yc * max(1, strheight(legend, units = \"user\", cex = cex)/yc)","    ychar <- yextra + ymax","    if (trace) ","        catn(\"  xchar=\", xchar, \"; (yextra,ychar)=\", c(yextra, ","            ychar))","    if (mfill) {","        xbox <- xc * 0.8","        ybox <- yc * 0.5","        dx.fill <- xbox","    }","    do.lines <- (!missing(lty) && (is.character(lty) || any(lty > ","        0))) || !missing(lwd)","    n.legpercol <- if (horiz) {","        if (ncol != 1) ","            warning(gettextf(\"horizontal specification overrides: Number of columns := %d\", ","                n.leg), domain = NA)","        ncol <- n.leg","        1","    }","    else ceiling(n.leg/ncol)","    has.pch <- !missing(pch) && length(pch) > 0","    if (do.lines) {","        x.off <- if (merge) ","            -0.7","        else 0","    }","    else if (merge) ","        warning(\"'merge = TRUE' has no effect when no line segments are drawn\")","    if (has.pch) {","        if (is.character(pch) && !is.na(pch[1L]) && nchar(pch[1L], ","            type = \"c\") > 1) {","            if (length(pch) > 1) ","                warning(\"not using pch[2..] since pch[1L] has multiple chars\")","            np <- nchar(pch[1L], type = \"c\")","            pch <- substr(rep.int(pch[1L], np), 1L:np, 1L:np)","        }","        if (!is.character(pch)) ","            pch <- as.integer(pch)","    }","    if (is.na(auto)) {","        if (xlog) ","            x <- log10(x)","        if (ylog) ","            y <- log10(y)","    }","    if (nx == 2) {","        x <- sort(x)","        y <- sort(y)","        left <- x[1L]","        top <- y[2L]","        w <- diff(x)","        h <- diff(y)","        w0 <- w/ncol","        x <- mean(x)","        y <- mean(y)","        if (missing(xjust)) ","            xjust <- 0.5","        if (missing(yjust)) ","            yjust <- 0.5","    }","    else {","        h <- (n.legpercol + (!is.null(title))) * ychar + yc","        w0 <- text.width + (x.intersp + 1) * xchar","        if (mfill) ","            w0 <- w0 + dx.fill","        if (do.lines) ","            w0 <- w0 + (seg.len + x.off) * xchar","        w <- ncol * w0 + 0.5 * xchar","        if (!is.null(title) && (abs(tw <- strwidth(title, units = \"user\", ","            cex = cex) + 0.5 * xchar)) > abs(w)) {","            xextra <- (tw - w)/2","            w <- tw","        }","        if (is.na(auto)) {","            left <- x - xjust * w","            top <- y + (1 - yjust) * h","        }","        else {","            usr <- par(\"usr\")","            inset <- rep_len(inset, 2)","            insetx <- inset[1L] * (usr[2L] - usr[1L])","            left <- switch(auto, bottomright = , topright = , ","                right = usr[2L] - w - insetx, bottomleft = , ","                left = , topleft = usr[1L] + insetx, bottom = , ","                top = , center = (usr[1L] + usr[2L] - w)/2)","            insety <- inset[2L] * (usr[4L] - usr[3L])","            top <- switch(auto, bottomright = , bottom = , bottomleft = usr[3L] + ","                h + insety, topleft = , top = , topright = usr[4L] - ","                insety, left = , right = , center = (usr[3L] + ","                usr[4L] + h)/2)","        }","    }","    if (plot && bty != \"n\") {","        if (trace) ","            catn(\"  rect2(\", left, \",\", top, \", w=\", w, \", h=\", ","                h, \", ...)\", sep = \"\")","        rect2(left, top, dx = w, dy = h, col = bg, density = NULL, ","            lwd = box.lwd, lty = box.lty, border = box.col)","    }","    xt <- left + xchar + xextra + (w0 * rep.int(0:(ncol - 1), ","        rep.int(n.legpercol, ncol)))[1L:n.leg]","    yt <- top - 0.5 * yextra - ymax - (rep.int(1L:n.legpercol, ","        ncol)[1L:n.leg] - 1 + (!is.null(title))) * ychar","    if (mfill) {","        if (plot) {","            if (!is.null(fill)) ","                fill <- rep_len(fill, n.leg)","            rect2(left = xt, top = yt + ybox/2, dx = xbox, dy = ybox, ","                col = fill, density = density, angle = angle, ","                border = border)","        }","        xt <- xt + dx.fill","    }","    if (plot && (has.pch || do.lines)) ","        col <- rep_len(col, n.leg)","    if (missing(lwd) || is.null(lwd)) ","        lwd <- par(\"lwd\")","    if (do.lines) {","        if (missing(lty) || is.null(lty)) ","            lty <- 1","        lty <- rep_len(lty, n.leg)","        lwd <- rep_len(lwd, n.leg)","        ok.l <- !is.na(lty) & (is.character(lty) | lty > 0) & ","            !is.na(lwd)","        if (trace) ","            catn(\"  segments2(\", xt[ok.l] + x.off * xchar, \",\", ","                yt[ok.l], \", dx=\", seg.len * xchar, \", dy=0, ...)\")","        if (plot) ","            segments2(xt[ok.l] + x.off * xchar, yt[ok.l], dx = seg.len * ","                xchar, dy = 0, lty = lty[ok.l], lwd = lwd[ok.l], ","                col = col[ok.l])","        xt <- xt + (seg.len + x.off) * xchar","    }","    if (has.pch) {","        pch <- rep_len(pch, n.leg)","        pt.bg <- rep_len(pt.bg, n.leg)","        pt.cex <- rep_len(pt.cex, n.leg)","        pt.lwd <- rep_len(pt.lwd, n.leg)","        ok <- !is.na(pch)","        if (!is.character(pch)) {","            ok <- ok & (pch >= 0 | pch <= -32)","        }","        else {","            ok <- ok & nzchar(pch)","        }","        x1 <- (if (merge && do.lines) ","            xt - (seg.len/2) * xchar","        else xt)[ok]","        y1 <- yt[ok]","        if (trace) ","            catn(\"  points2(\", x1, \",\", y1, \", pch=\", pch[ok], ","                \", ...)\")","        if (plot) ","            points2(x1, y1, pch = pch[ok], col = col[ok], cex = pt.cex[ok], ","                bg = pt.bg[ok], lwd = pt.lwd[ok])","    }","    xt <- xt + x.intersp * xchar","    if (plot) {","        if (!is.null(title)) ","            text2(left + w * title.adj, top - ymax, labels = title, ","                adj = c(title.adj, 0), cex = cex, col = title.col)","        text2(xt, yt, labels = legend, adj = adj, cex = cex, ","            col = text.col, font = text.font)","    }","    invisible(list(rect = list(w = w, h = h, left = left, top = top), ","        text = list(x = xt, y = yt)))","}"]}},"theme":{"colors":["#F44336","#2196F3","#4CAF50","#FFEB3B","#9C27B0","#03A9F4","#8BC34A","#FF9800","#3F51B5","#E91E63","#CDDC39","#FF5722"],"chart":{"style":{"fontFamily":"Roboto","color":"#212121"}},"xAxis":{"gridLineWidth":1,"gridLineColor":"#F3F3F3","lineColor":"#F3F3F3","minorGridLineColor":"#F3F3F3","tickColor":"#F3F3F3","tickWidth":1},"yAxis":{"gridLineColor":"#F3F3F3","lineColor":"#F3F3F3","minorGridLineColor":"#F3F3F3","tickColor":"#F3F3F3","tickWidth":1},"plotOptions":{"boxplot":{"fillColor":"#FAFAFA"}},"legendBackgroundColor":"rgba(0, 0, 0, 0.5)","background2":"#505053","dataLabelsColor":"#B0B0B3","textColor":"#C0C0C0","contrastTextColor":"#F0F0F3","maskColor":"rgba(255,255,255,0.3)"},"conf_opts":{"global":{"Date":null,"VMLRadialGradientURL":"http =//code.highcharts.com/list(version)/gfx/vml-radial-gradient.png","canvasToolsURL":"http =//code.highcharts.com/list(version)/modules/canvas-tools.js","getTimezoneOffset":null,"timezoneOffset":0,"useUTC":true},"lang":{"contextButtonTitle":"Chart context menu","decimalPoint":".","downloadJPEG":"Download JPEG image","downloadPDF":"Download PDF document","downloadPNG":"Download PNG image","downloadSVG":"Download SVG vector image","drillUpText":"Back to {series.name}","invalidDate":null,"loading":"Loading...","months":["January","February","March","April","May","June","July","August","September","October","November","December"],"noData":"No data to display","numericSymbols":["k","M","G","T","P","E"],"printChart":"Print chart","resetZoom":"Reset zoom","resetZoomTitle":"Reset zoom level 1:1","shortMonths":["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"thousandsSep":" ","weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"]}},"type":"chart","fonts":"Roboto","debug":false},"evals":[],"jsHooks":[]}</script>
<!-- ## Pairwise Comparisons -->
<!-- braucht geneList -->
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="sample-comparison.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="scatter-plot-matrices.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
