![CRISPRAnalyzeR](https://github.com/boutroslab/crispr-analyzer/blob/master/images/CRISPRAnalyzR_logo5.png)
# CRISPRAnalyzeR - Fully-interactive Analysis and Documentation of Pooled CRISPR Screens
Welcome to the CRISPRAnalyzeR Github page!

CRISPRAnalyeR is a web-based, fully interactive suite for the analysis and documentation of pooled CRISPR Screens.

#### See the live-demo of [CRISPRAnalyzeR](http://crispr-analyzer.dkfz.de)

CRISPRAnalyzeR has been specifically developed to provide a fully-interactive, hollistic and exploratory analysis of pooled CRISPR Screens especially for those people that perform the screens themselves.
You can easily analyse your screen using 6 different hit calling methods, 2 essential gene calling methods as well as perform gene annotation anrichment and gene set analysis - all in a convenient web-browser interface.

**All you need is your sequencing data and the pooled CRISPR screen library file (we provide you with the most common ones) - and CRISPRAnalyzeR will help you to go from rawdata to potential followup candidates!**

<img src="https://github.com/boutroslab/crispr-analyzer/blob/master/images/CRISPRAnalyzer_screen_scheme.png" alt="alt text" width="50%" style="align:center;" >


## What makes CRISPRAnalyzeR the perfect tool to analyze pooled CRISPR Screens?

CRISPRAnalyzeR uses what we call a **guided-analysis** approach. This means you will be **guided through the whole analysis, focusing on the most important things - your data**.

In brief, CRISPAnalyzeR consists of **four sections**:
- Screen Quality Estimation
- Hit Calling using multiple published hit calling methods
- Hit Confirmation with extensive Gene information, sgRNA information, Gene Set Analysis and much more
- Download of the interactive report that provides a comprehensive documentation of your screen

<img src="https://github.com/boutroslab/crispr-analyzer/blob/master/images/CRISPRAnalyzeR_5-columns_small.png" alt="alt text" width="50%" style="align:center;" >

### The principle CRISPRAnalyzeR Guided-Analysis Workflow
<img src="https://github.com/boutroslab/crispr-analyzer/blob/master/images/workflow3.png" alt="alt text" width="50%" style="align:center;" >



**Analysing CRISPR Screens has never been easier - and has never been so much fun!**


## How to download the CRISPRAnalyzeR

#### Please check our [live demo](http://crispr-analyzer.dkfz.de), which you can also use to analyse your screening data.

You can get the CRISPRAnalyzeR suite as source code or as a ready-to-use Docker Container.

#### Minimum System Requirements
CRISPRAnalyzeR is based on R Shiny-Server and uses many different R packages and tools.
For a source code installation, we recommend the use of Ubuntu.

 | Source Code Installation | Docker Container Installation
----|---|----
Operating System|Ubuntu|Any supported OS by Docker
CPU|Dual Core (Quad Core recommended)| Dual Core (Quad Core recommended)
RAM|8 GB| 8GB
HDD|512 GB (SSD recommended)| 512 GB (SSD recommended)
Additional Software Packages | **See list below!** | All included in container


### Source Code


### Ready-to-use Docker Container

## What CRISPRANalyzeR offers you

## How to install CRISPRAnalyzeR using the provided Source Code

## How to install CRISPRAnalyzeR using the provided Docker Container

## How to perform an Analysis using CRISPRAnalyzeR
*YOUTUBE VID HERE*

## Pre-made sgRNA library FASTA files

## Common settings for published CRISPR libraries
