<!-- 
save as 'screenQuality.Rmd'
child of 'report.Rmd'
--> 



<!--FastQ Quality Plots-->
```{r}
fqQCbool <- if(extractedFiles$rqc == "") FALSE else TRUE
```

# Data Review

```{r data-overview, results='asis'}


top <- "<div class='section'><div class='container'><div class='row'><div class='col-md-12'><table class='table'>
<thead>
<tr>
<th>Filename Original</th>
<th>Provided Name</th>
<th>Type</th>
<th>Size</th>
<th>sgRNA</br>Extraction Ratio</th>
<th>Mapped</br>to Reference</th>
<th>Passed</br>Quality Threshold</th>
<th>Used</br>in Total</th>
</tr>
</thead>
<tbody>"
  
  bottom <- "</tbody></table></div></div></div></div>"
  
  fileinfo <- data.frame("NameOriginal" = "",
                         "NameNew" = "",
                         "Type" = "",
                         "Size" = "",
                         "extractRatio" = "not available",
                         "bt2mapping" = "not available",
                         "bt2matching" = "not available",
    stringsAsFactors = FALSE
  )
  # loop through files and give back rdy to use HTML

  for(i in 1:length(extractedFiles$paths))
  {
    # Set initals
    fileinfo[i,"extractRatio"] <- "not available"
    fileinfo[i,"bt2mapping"] <- "not available"
    fileinfo[i,"bt2matching"] <- "not available"
    
    # check if paths are different, if it is it is a readcount file as well as a FASTQ file available
    if(extractedFiles$paths[i] == extractedFiles$oldpaths[i])
    {
      fileinfo[i,"Type"] <- "Readcount"
    }
    else # it is a fastq files uploaded, so we need to change things a bit
    {
      fileinfo[i,"Type"] <- "Fastq"
      # get extractRatio
      #fileinfo[i,"extractRatio"] <- paste((as.numeric(seqFiles()$extractRatio[i])), "%", sep="")
      
      fileinfo[i,"extractRatio"] <- paste((as.numeric(extractedFiles$extractRatio[[i]])), "%", sep="")
      
      
      
      # get bt2mapping
      if(!is.na(extractedFiles$bt2mapping[[i]][1]))
      {
        #fileinfo[i,"extractRatio"] <- extractedFilesbt2mapping[i]
        mappinginfo <- c(paste("Not aligned:", extractedFiles$bt2mapping[[i]][["notalign"]], sep= " "), paste("Aligned once:", extractedFiles$bt2mapping[[i]][["singlealign"]], sep= " "), paste("Aligned multiple times:", extractedFiles$bt2mapping[[i]][["multiplealign"]], sep= " "))
        mapping <- paste(mappinginfo, collapse = "<br/>")
        
        fileinfo[i,"bt2mapping"] <- mapping
      }
      
      # get bt2matching
      if(!is.na(extractedFiles$bt2matching[[i]][1]))
      {
        #fileinfo[i,"extractRatio"] <- extractedFilesbt2mapping[i]
        matchinginfo <- extractedFiles$bt2matching[[i]]
        
        fileinfo[i,"bt2matching"] <- paste(matchinginfo, "%", sep = "")
      }
      
      # final output
      if(is.na(extractedFiles$extractRatio[[i]]) || is.na(extractedFiles$bt2mapping[[i]]) || is.na(extractedFiles$bt2matching[[i]]))
      {
        fileinfo[i,"finaloutput"] <- "not available"
      } else {
        fileinfo[i,"finaloutput"] <- paste(round( ( (as.numeric(extractedFiles$extractRatio[[i]])/100) * (as.numeric(gsub(pattern = "%", replacement = "", x = extractedFiles$bt2mapping[[i]][["singlealign"]]))/100) * (as.numeric(extractedFiles$bt2matching[[i]]) /100 )  ) * 100, digits = 2),"%", sep="")
      }
      
      
    }
    # get Size of file
    fileinfo[i,"Size"] <-  paste(extractedFiles$filesize[[i]],"MegaByte", sep=" ")
    
    # Uploaded name
    fileinfo[i,"NameOriginal"] <- extractedFiles$names[i]
    
    # New name
    fileinfo[i,"NameNew"] <- extractedFiles$gen_names[i]
    
    
    # make HTML
    if(i==1)
    {
      fileinfohtml <- paste("<tr><td>",fileinfo[i,"NameOriginal"],"</td><td>",fileinfo[i,"NameNew"],"</td><td>",fileinfo[i,"Type"],"</td><td>",fileinfo[i,"Size"],"</td><td>", fileinfo[i,"extractRatio"],"</td><td>", fileinfo[i,"bt2mapping"],"</td><td>", fileinfo[i,"bt2matching"],"</td><td>", fileinfo[i,"finaloutput"],"</td></tr>", sep="")
    } else
    {
      fileinfohtml <- c(fileinfohtml, paste("<tr><td>",fileinfo[i,"NameOriginal"],"</td><td>",fileinfo[i,"NameNew"],"</td><td>",fileinfo[i,"Type"],"</td><td>",fileinfo[i,"Size"],"</td><td>", fileinfo[i,"extractRatio"],"</td><td>", fileinfo[i,"bt2mapping"],"</td><td>", fileinfo[i,"bt2matching"],"</td><td>", fileinfo[i,"finaloutput"],"</td></tr>", sep="") )
    }
  }

  outputhtml <- paste(fileinfohtml, sep ="", collapse= " ")
  outputhtml <- paste(top, outputhtml, bottom, sep="", collapse = " ")
  
  cat(outputhtml)
  




# 
# 
#   top <- "<div class='section'>
#       <div class='container'>
#   <div class='row'>
#   <div class='col-md-12'>
#   <h4 class='text-success text-left'>
#   <i class='fa fa-files-o fa-fw'></i>Uploaded / Generated Rawdata Files</h4>
#   <table class='table'>
#   <thead>
#   <tr>
#   <th>Filename Original</th>
#   <th>Provided Name</th>
#   <th>Type</th>
#   <th>sgRNA</br>Extraction Ratio</th>
#   <th>Reads</br>Mapped to Reference</th>
#   <th>Reads</br>Passed Quality Threshold</th>
#   <th>Reads</br>Used in Total</th>
#   </tr>
#   </thead>
#   <tbody>"
#   
#   bottom <- "</tbody>
#                       </table>
#   
#   </div>
#   </div>
#   </div>
#   </div>"
#   
#   fileinfo <- data.frame("NameOriginal" = "",
#                          "NameNew" = "",
#                          "Type" = "",
#                          "Size" = "",
#                          "extractRatio" = "not available",
#                          "bt2mapping" = "not available",
#                          "bt2matching" = "not available",
#     stringsAsFactors = FALSE
#   )
#   # loop through files and give back rdy to use HTML
#   
#   for(i in 1:length(extractedFiles$paths))
#   {
#     # Set initals
#     fileinfo[i,"extractRatio"] <- "not available"
#     fileinfo[i,"bt2mapping"] <- "not available"
#     fileinfo[i,"bt2matching"] <- "not available"
#     
#     # check if paths are different, if it is it is a readcount file as well as a FASTQ file available
#     if(extractedFiles$paths[i] == extractedFiles$oldpaths[i])
#     {
#       fileinfo[i,"Type"] <- "Readcount"
#     }
#     else # it is a fastq files uploaded, so we need to change things a bit
#     {
#       fileinfo[i,"Type"] <- "Fastq"
#       # get extractRatio
#       #fileinfo[i,"extractRatio"] <- paste((as.numeric(seqFiles$extractRatio[i])), "%", sep="")
#       
#       fileinfo[i,"extractRatio"] <- paste((as.numeric(extractedFiles$extractRatio[[i]])), "%", sep="")
#       
#       # get bt2mapping
#       if(!is.na(extractedFiles$bt2mapping[[i]][1]))
#       {
#         #fileinfo[i,"extractRatio"] <- extractedFilesbt2mapping[i]
#         mappinginfo <- c(paste("Not aligned:", extractedFiles$bt2mapping[[i]][["notalign"]], sep= " "), paste("Aligned once:", extractedFiles$bt2mapping[[i]][["singlealign"]], sep= " "), paste("Aligned multiple times:", extractedFiles$bt2mapping[[i]][["multiplealign"]], sep= " "))
#         mapping <- paste(mappinginfo, collapse = "<br/>")
#         
#         fileinfo[i,"bt2mapping"] <- mapping
#       }
#       
#       # get bt2matching
#       if(!is.na(extractedFiles$bt2matching[[i]][1]))
#       {
#         #fileinfo[i,"extractRatio"] <- extractedFilesbt2mapping[i]
#         matchinginfo <- extractedFiles$bt2matching[[i]]
#         
#         fileinfo[i,"bt2matching"] <- paste(matchinginfo, "%", sep = "")
#       }
#       
#       # final output
#       if(is.na(extractedFiles$extractRatio[[i]]) || is.na(extractedFiles$bt2mapping[[i]]) || is.na(extractedFiles$bt2matching[[i]]))
#       {
#         fileinfo[i,"finaloutput"] <- "not available"
#       } else {
#         fileinfo[i,"finaloutput"] <- paste(round( ( (as.numeric(extractedFiles$extractRatio[[i]])/100) * (as.numeric(gsub(pattern = "%", replacement = "", x = extractedFiles$bt2mapping[[i]][["singlealign"]]))/100) * (as.numeric(extractedFiles$bt2matching[[i]]) /100 )  ) * 100, digits = 2),"%", sep="")
#       }
#       
#       
#     }
#     # get Size of file
#     fileinfo[i,"Size"] <-  paste(extractedFiles$filesize[[i]],"MegaByte", sep=" ")
#     
#     # Uploaded name
#     fileinfo[i,"NameOriginal"] <- extractedFiles$names[i]
#     
#     # New name
#     fileinfo[i,"NameNew"] <- extractedFiles$gen_names[i]
#     
#     
#     # make HTML
#     if(i==1)
#     {
#       fileinfohtml <- paste("<tr>
#                       <td>",fileinfo[i,"NameOriginal"],"</td>
#       <td>",fileinfo[i,"NameNew"],"</td>
#       <td>",fileinfo[i,"Type"],"</td>
#       <td>
#       ",fileinfo[i,"Size"],"
#       </td>
#       <td>
#       ", fileinfo[i,"extractRatio"],"
#       </td>
#       <td>
#       ", fileinfo[i,"bt2mapping"],"
#       </td>
#       <td>
#       ", fileinfo[i,"bt2matching"],"
#        </td>
#       <td>
#       ", fileinfo[i,"finaloutput"],"
#       </td>
#       </tr>", sep="")
#     } else
#     {
#       fileinfohtml <- c(fileinfohtml, paste("<tr>
#                       <td>",fileinfo[i,"NameOriginal"],"</td>
#                            <td>",fileinfo[i,"NameNew"],"</td>
#                            <td>",fileinfo[i,"Type"],"</td>
#                            
#                             <td>
#                             ", fileinfo[i,"extractRatio"],"
#                             </td>
#                             <td>
#                             ", fileinfo[i,"bt2mapping"],"
#                             </td>
#                             <td>
#                             ", fileinfo[i,"bt2matching"],"
#                              </td>
#                             <td>
#                             ", fileinfo[i,"finaloutput"],"
#                              </td>
#                            </tr>", sep="") )
#     }
#   }
#   lengthfiles <- nrow(fileinfo)
#   # Add Library
#   fileinfo[lengthfiles+1,"Type"] <- "Library File"
#   fileinfo[lengthfiles+1,"Size"] <-  paste(round((file.size(extractedFiles$libPath)/1024)/1024, digits = 2),"MegaByte", sep=" ")
#   fileinfo[i,"NameOriginal"] <- extractedFiles$libName
#   fileinfo[i,"NameNew"] <- extractedFiles$libName
#   
#   outputhtml <- paste(fileinfohtml, sep ="", collapse= " ")
#   outputhtml <- paste(top, outputhtml, bottom, sep="", collapse = " ")
# 
#   cat(outputhtml)
  
```

```{r, eval=fqQCbool, child='sqFastQqcChild.Rmd'}

```













<!--Readcount Plots-->

# Screen Readcount

This section is a screen overview with descriptive statistics in raw and normalized data.
Normalization is done with a method from the [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) package that moderates dispersion.

## Overview

Below are descriptive statistics about sgRNA and gene readcounts.

### sgRNA Readcount Statistics

```{r}
df <- results$statsGeneral$basic$sgRNA
tryFunction(Table_small(df, filename = "sgRNA_basic_stats")   , place = 'sq')
```

### Gene Readcount Statistics

```{r}
df <- results$statsGeneral$basic$Gene
tryFunction(Table_small(df, filename = "Gene_basic_stats")   , place = 'sq')
```

## Readcount Distribution

In general readcounts are expected to follow an over-dispersed poisson distribution.
Depending on your experiment you might expect the overall readcount of certain samples to be lower than others.
In addition, several effects -- like drop outs for example -- can manifest as distinct peaks or outliers.

### Pseudodensity

These pseudodensity curves are based on normalized, log2-transformed readcount histograms for each sample.
(You can drag a rectangle to zoom in)

```{r sq_readcountsDistro}
tryFunction(Plot_coverage_readDistro( results$readDistribution, bApp = FALSE )   , place = 'sq')
```

### Essential and Non-Essential Density

```{r sq_essentialDensity1}
tryFunction(Plot_essential_distribution(results$essentialDistribution[[1]], bApp = TRUE, filename = "QualityControl_Essential_Distribution_Untreated")  , place = 'sq')
```

```{r sq_essentialDensity2}
tryFunction(Plot_essential_distribution(results$essentialDistribution[[2]], bApp = TRUE, filename = "QualityControl_Essential_Distribution_Treated")  , place = 'sq')
```

### Cumulative Frequency for Gene Read Counts

Below you find the log2-transformed, normalized cumulative frequency for each sample's gene read counts.

```{r}
tryFunction(
  Plot_CDF(data = results$CDF_list[["gene"]], readtype = "gene", filenames = extractedFiles$gen_names, title = "Cumulative Frequency", subtitle = "based on Gene Read Counts", filename="QualityControl_ReadCount_CDF_Gene"),
  place = "sq"
)
```

### Cumulative Frequency for sgRNA Read Counts

Below you find the log2-transformed, normalized cumulative frequency for each sample's sgRNA read counts.

```{r}
tryFunction(
  Plot_CDF(data = results$CDF_list[["sgRNA"]], readtype = "sgrna", filenames = extractedFiles$gen_names, title = "Cumulative Frequency", subtitle = "based on sgRNA Read Counts", filename="QualityControl_ReadCount_CDF_sgRNA"),
  place = "sq"
)
```



### Boxplots

These are the normalized, log2-transformed read count distributions.
(Click on the Legend to exclude/include outliers)

```{r sq_readcountsDistroBoxGene}
tryFunction(Plot_coverage_readDistroBox( results$readDistributionBoxNorm, bApp = FALSE , type = "Gene")   , place = 'sq')
```




```{r, results='asis'}
posBool <- as.logical(results$ctrls$pos)
negBool <- as.logical(results$ctrls$non)
if( any(c(posBool, negBool)) ) cat("## Controls\n\n")
cat("Below are descriptive statistics about gene readcounts of genes defined as control for each sample.\n\n")
```

<!--Controls-->
```{r, results='asis', eval=posBool}
cat("### Positive Controls\n\n")
```
```{r, eval=posBool}
df <- results$statsGeneral$ctrl.p
tryFunction(Table_small( df, type = "ctrl" , filename ="sqStats_ctrl_pos")   , place = 'sq')
```

```{r, results='asis', eval=negBool}
cat("### Non-targeting Controls\n\n")
```
```{r, eval=negBool}
df <- results$statsGeneral$ctrl.n
tryFunction(Table_small( df, type = "ctrl", filename ="sqStats_ctrl_pos")   , place = 'sq')
```


## Readcount Depth

This plot shows the readcount of each gene divided by the number of sgRNAs that contributed to it.
Spikes indicate a higher readcount per sgRNA for this particular gene. One would expect no outstanding spikes within the untreated datasets
However, spikes within the treated datasets indicate a read count enrichment per sgRNA for this particular gene. 
positive and non-targeting controls are highlighted in red and blue respectively.


```{r sq_readcountDepth_eval, include=FALSE, message=FALSE}
# pro coding inc...
# I can't loop in a chunk if I want highcharter plots to render correctly
# here I just create the plot 20 times
# every chunk is checked whether it should be evaluated
# a bool arr is used for checking
# only problem: if more than 20 needed, only 20 will be rendered
datasets <- c(results$compare[[1]], results$compare[[2]])
datasetBool <- rep(FALSE, 20)
datasetBool[1:length(datasets)] <- TRUE
counter <- 1
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_readDepth( results$readDepth, dataset, results$ctrls$non, results$ctrls$pos, bApp = FALSE )   , place = 'sq')
```




















## Readcounts

This table contains the readcoutns of each sgRNA in each sample.
This is the basis for all plots and statistics in this report.
As an overview for each sample tables with descriptive statistics about each gene in each sample are attached.

```{r sq_readcountsGenes}
df <- results$readcount
tryFunction(Table_big(df)   , place = 'sq')
```

```{r sq_readcountsTable_eval, include=FALSE, message=FALSE}
# pro coding inc...
# I can't loop in a chunk if I want highcharter plots to render correctly
# here I just create the plot 20 times
# every chunk is checked whether it should be evaluated
# a bool arr is used for checking
# only problem: if more than 20 needed, only 20 will be rendered
datasets <- c(results$compare[[1]], results$compare[[2]])
datasetBool <- rep(FALSE, 20)
datasetBool[1:length(datasets)] <- TRUE
counter <- 1
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Table_big(results$statsGeneral$dataset, dataset)   , place = 'sq')
```















<!--Unmapped-->

# sgRNA Coverage

Missing sgRNAs are the number of sgRNAs which were defined in the sgRNA library but are not present in the dataset.
Missing genes are the number of genes for which all sgRNAs were not present in the dataset.
Depending on the experiment certain sgRNAs or genes might be expected to be missing.

## Missing sgRNAs and Genes

### Missing Genes

This number of genes did not result in any read count and are therefore considered missing in the respective sample.

```{r sq_coverageGenes}
tryFunction(Plot_coverage_unmapped( results$unmappedGenes, "gene", bApp = FALSE )   , place = 'sq')
```

### Missing sgRNAs

This number of sgRNAs did not result in read counts and are therefore considered missing in the respective sample.

```{r sq_coverageSgRNAs}
tryFunction(Plot_coverage_unmapped( results$unmappedGenes, "sgRNA", bApp = FALSE )   , place = 'sq')
```

### \% Missing Non-targeting sgRNAs

This is the percentage of missing sgRNAs among the genes specified as non-targeting controls.

```{r sq_coverageNonTsgRNA}
if( results$ctrls$non == FALSE ){
    Plot_blank("hc", msg = "No non-targeting controls were defined.")
} else {
tryFunction(Plot_coverage_unmapped( results$unmappedGenes, "non", bApp = FALSE )   , place = 'sq')
}
```

### \% Missing Postive sgRNAs

This is the percentage of missing sgRNAs among the genes specified as positive controls.

```{r sq_coveragePosSgRNA}
if( results$ctrls$pos == FALSE ){
  Plot_blank("hc", msg = "No positive controls were defined.")
} else {
tryFunction(Plot_coverage_unmapped( results$unmappedGenes, "pos", bApp = FALSE )   , place = 'sq')
}
```

## Designs per Gene 

These plots provide and overview of the representation of sgRNAs per gene within your data.
In case every gene is covered by all sgRNAs, a single column at 100\% is expected.
Moreover, genes with a low percentage of present sgRNAs might also show a reduced read count. 

```{r sq_designsGene_eval, include=FALSE, message=FALSE}
# pro coding inc...
# I can't loop in a chunk if I want highcharter plots to render correctly
# here I just create the plot 20 times
# every chunk is checked whether it should be evaluated
# a bool arr is used for checking
# only problem: if more than 20 needed, only 20 will be rendered
datasets <- c(results$compare[[1]], results$compare[[2]])
datasetBool <- rep(FALSE, 20)
datasetBool[1:length(datasets)] <- TRUE
counter <- 1
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```

```{r, results='asis', eval=datasetBool[counter]}
cat("###", datasets[counter])
```
```{r, eval=datasetBool[counter]}
dataset <- datasets[counter]
counter <- counter + 1
tryFunction(Plot_coverage_geneDesigns(results$geneDesigns, dataset, nBreaks = 40, bApp = FALSE )   , place = 'sq')
```































# Sample Comparison

In this chapter single samples are comapred to visualize replicate quality and effects in different samples.

## Overview

Normalized readcount foldchanges or Z-Ratios between all samples on both gene and sgRNA level are shown in this section.

### Foldchange of Gene Effects

```{r, message=FALSE, warning=FALSE}
tryFunction(highcharts.sampleComp(results$sampleList, sample = NULL, type = "genes", plot = "boxplot", 
                      sorted = TRUE, info = "log2fc" , theme = TRUE, addline=FALSE)   , place = 'sq')
```

### Z-Ratio of Gene Effects

```{r, message=FALSE, warning=FALSE}
tryFunction(highcharts.sampleComp(results$sampleList, sample = NULL, type = "genes", plot = "boxplot", 
                      sorted = TRUE, info = "zratio" , theme = TRUE, addline=FALSE)   , place = 'sq')
```

### Foldchange of sgRNA Effects

```{r, message=FALSE, warning=FALSE}
tryFunction(highcharts.sampleComp(results$sampleList, sample = NULL, type = "sgrna", plot = "boxplot", 
                      sorted = TRUE, info = "log2fc" , theme = TRUE, addline=FALSE)   , place = 'sq')
```










<!-- ## Pairwise Comparisons -->
<!-- braucht geneList -->















```{r sq_splom_eval, include=FALSE, message=FALSE}
# pro coding inc...
# I can't loop in a chunk if I want highcharter plots to render correctly
# here I just create the plot 20 times
# every chunk is checked whether it should be evaluated
# a bool arr is used for checking
# only problem: if more than 20 needed, only 20 will be rendered
nPlots <- geneList$SPLOM
plotsBool <- rep(FALSE, 20)
if(length(nPlots) > 0) plotsBool[1:length(nPlots)] <- TRUE
counter <- 1
```

```{r, results='asis', eval=plotsBool[counter]}
inf <- geneList$SPLOM[[counter]]
cat("## Scatter Plot Matrices\n\n")
cat("Scatter plot matrices, which were saved by the user, are provided below.\n\n")
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("\n\n### Scatter Plot Matrix", counter, "\n\n")
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts with positive and non-targeting controls highlighted in red and blue respectively.\n"))
if(inf$blog) cat(" Axis are log10 transformed.")
cat("\n\n\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS
tryFunction(Plot_Replicates_SPLOM( data, inf$dsNames, bLog = inf$blog, bAggro = inf$baggro, cCtrl = inf$cctrl )   , place = 'sq')
```





















```{r sq_scatter_eval, include=FALSE, message=FALSE}
nPlots <- geneList$replicates
plotsBool <- rep(FALSE, 20)
if(length(nPlots) > 0) plotsBool[1:length(nPlots)] <- TRUE
counter <- 1
```

```{r, results='asis', eval=plotsBool[counter]}
cat("## Single Scatter Plots\n\n")
cat("These are scatter plots, saved by the user.\n\n")
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```

```{r, results='asis', eval=plotsBool[counter]}
cat("### Scatterplot", counter, "\n\n")
inf <- nPlots[[counter]]
cat(paste("Shown are", if(inf$baggro) "gene" else "sgRNA", "readcounts of", inf$YY, "over", inf$XX,
          "on a", if(inf$blog) "log10" else "linear", "scale.\n"))
cat("Controls and user defined genes are highlighted.\n\n")
```
```{r, eval=plotsBool[counter]}
counter <- counter + 1
data <- results$readCountVS 
tryFunction(Plot_replicates_pair(inf$XX, inf$YY, data, inf$ctrl, bLog = inf$blog, aggro = inf$baggro, bApp = FALSE,
                     filename = paste("QualityControl_Replicates_", inf$XX, "_VS_", inf$YY, sep=""), labelgene = inf$highlighted, results = results$compare)   , place = 'sq')
```
























<!--PCA plot-->

# PCA

Sometimes a principal component analysis to visualize high-dimensional data.
Here, genes are represented as data points with their score in each principal component.
Vectors show the coefficients of each sample to the principal component.
Controls that are supposed to show an effect, should be separable from the data crowd on the first principal component.
Vectors of replicates should point in a similar direction.

```{r sq_pca}
tryFunction(Plot_pca(pca = results$pca$genes, title = "Principal Component Analysis", 
         subtitle = NULL, export = TRUE, legend = TRUE, scale=8)   , place = 'sq')
```







<!--Heatmap-->

```{r, include=FALSE, message=FALSE}
headderBool <- length(geneList$heatmap) > 0
```
```{r, results='asis', eval=headderBool}
cat('# Heatmaps\n\n')
```

```{r, eval=headderBool}
HTML("<div><p>This section contains all heatmaps the user created during the analysis.</div></p>")
```

```{r sq_heatmap_eval, include=FALSE, message=FALSE}
nPlots <- geneList$heatmap
plotsBool <- rep(FALSE, 20)
if(length(nPlots) > 0) plotsBool[1:length(nPlots)] <- TRUE
counter <- 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

```{r, results='asis', eval=plotsBool[counter]}
show <- geneList$heatmap[[counter]]$show
type <- geneList$heatmap[[counter]]$type
checks <- geneList$heatmap[[counter]]$checks
tp <- if(type == "enriched") "within the top 5% of reads." else "within the lowest 5% of reads."
txt <- switch (show,
  geneabundance = "is the abundance of readcounts summed up for each gene compared to all reads in the dataset.",
  genereadcount = "are readcounts summed up for each gene.",
  genethreshold = paste("is the percentage of sgRNA for each gene having a readcount", tp),
  sgrnaabundance = "is the abundance of a single sgRNA compared to all reads in the dataset.",
  sgrnareadcount = "are sgRNA readcounts.",
  unmapped = "the percentage of missing sgRNAs per Gene."
)
cat("## Heatmap:", show, sprintf("(%s)", counter), "\n\n")
cat("Shown", txt)
if("log" %in% checks) cat(" Colors represent the decadic logarithm of values shown.")
cat("\n(The plot is still interactive and can be zoomed)")
```
```{r, eval=plotsBool[counter], out.width='100%'}
df <- geneList$heatmap[[counter]]$plot
df
```
```{r, include=FALSE}
counter <- counter + 1
```

